---
title: "Both datasets exploratory"
output: 
  pdf_document:
    toc: yes
  html_document:
    toc: yes
    theme: simplex
---

# Objective

The idea of this document is to draft what we need to do to integrate the GSEA pathway analysis with the diff. expression analysis we did with DESeq2. We propose the following pipeline:

1. We fit a model with a formula of the type ` Count_ij ~ Treat_i + Cell_j + Interaction_ij`

2. For every gene, test an hypothesis if there is a treatment effect in the cell specific expression.

3. For every gene, summarize the TPM level of each treatment, cell and interaction.

4. For every cell, take the genes that are differentially expressed and are among the top $K$ most expressed genes for both treatments (the assumption here is to avoid cases where the `log2FC` value is extreme due to a very low quantity of reads in either treatment).

5. Using the t.stat values as a signal-2-noise metric, do a pathway analysis with GSEA.

```{r load,include=FALSE,echo=FALSE,eval=TRUE}
library(tidyverse)
library(here)
library(viridis)
library(ggtech)
library(hexbin)
library(gridExtra)
library(forcats)
library(scales)
library(knitr)
library(ggrepel)
library(ggjoy)
library(clusterProfiler)
pal = viridis(1e3, option = "D")

## parameters
thr = .5
fdr = .01

theme_set(theme_minimal() + 
            theme(legend.position = "top"))

knitr::opts_chunk$set(eval = TRUE,
                      echo = FALSE,
                    include = TRUE,
                    warning = FALSE,
                    message = FALSE)

load(here("apps/Fig3_dashboad/fig3data.RData"))

diff_genes = tribble(
  ~ cell , ~ results,
  "EBV", diff_genes_EBV,
  "NOKS", diff_genes_NOKS
)

DE_results = diff_genes %>% 
  mutate(batch = "old")

DE_rsem = rsem_data %>% 
  mutate(
    rep = str_split(file,"rep") %>% 
      map_chr(~ .[2]) %>% 
      map_int(as.integer)
  )


load(here("apps/Fig5_dashboard/fig5data.RData"))

DE_results = bind_rows(
  DE_results,
  diff_genes %>% 
    dplyr::select(-contrast) %>% 
    mutate(batch = "new")
)

DE_rsem = bind_rows(
  DE_rsem %>% mutate(batch = "old"),
  rsem_data %>% mutate(batch = "new")
) %>% 
  mutate(
    treatment = fct_collapse(treatment,mc = c("methyl","MC"))
  )

rm(diff_genes_EBV,diff_genes_NOKS,rsem_data,alignment,rlogmat,diff_genes)

```

# TPM summaries

Both data batches are grouped, so we are going to rank the genes by the mean TPM.

```{r tpm_summary,cache=TRUE, include=FALSE}

DE_TPM = DE_rsem %>% 
  unnest() %>% 
  group_by(cell,treatment,batch,gene_id) %>% 
  summarize_at(
    vars(one_of("TPM")),
    funs( mean = mean(.))) %>% 
  ungroup() %>% 
  gather(metric, value,-(1:4)) %>% 
  nest(-batch,-cell,.key = "tpm.rank") %>% 
  mutate(
    tpm.rank = map(tpm.rank,spread,treatment,value) %>% 
      map( mutate_if,is.numeric,funs(perc = percent_rank(desc(.))))
  )

```


```{r tpm_rank, fig.width=7.5,fig.height=5,message=FALSE}


DE_TPM %>% 
  unnest() %>% 
  ggplot(aes(none_perc,mc_perc))+
  stat_binhex(bins = 50)+
  facet_grid(batch ~ cell)+
  scale_fill_gradientn(colours = pal,trans = "log10",
                         labels = trans_format('log10',math_format(10^.x)),
                         guide = guide_colourbar(title = NULL,
                           barheight = unit(0.01,"npc"),
                           barwidth = unit(0.9,"npc")))+
  ylab("rank: mc treatment")+xlab("rank: no treatment")+
  geom_rect(xmin = .75,xmax = 1,ymin = 0,ymax = .25,
            linetype = 2,colour = "red",alpha = 0)+
  geom_abline(intercept = c(thr,-thr),slope = c(1,1),colour = "orange",
              linetype = 2)

  
  
```

We further explore the genes in the red-square there are the genes that go from being unexpressed without treatment to be very expressed with treatment. In total, there are:


```{r red_table}

DE_TPM %>% 
  unnest() %>% 
  filter(none_perc >= .75 & mc_perc <= .25) %>% 
  group_by(batch , cell ) %>% 
  summarize(
    genes_in_square  = n()
  ) %>% 
  kable(digits = 0)

```

```{r red_square_plot,fig.width=7,fig.height=4 }

DE_TPM %>% 
  unnest() %>% 
  filter(none_perc >= .75 & mc_perc <= .25) %>% 
  dplyr::select(cell,batch,gene_id,mc,none) %>% 
  gather(treatment,mean_tpm,-(1:3)) %>% 
  separate(gene_id,c("ensembl","symbol"),sep = "\\_") %>% 
  mutate(batch = factor(batch,levels = c("old","new")),
         treatment = factor(treatment,levels = c("none","mc"))) %>% 
  ggplot(aes(treatment,mean_tpm))+
  geom_line(aes(group=symbol),size = .5,colour = "grey")+
  geom_point(aes(colour = cell))+
  scale_color_tech("google")+
  facet_wrap( ~ batch + cell,nrow = 1)+
  geom_hline(yintercept = 100,colour = "red",linetype = 2)+
  coord_cartesian(ylim = c(0,500))

```

In the figure above, it is shown that as expected when no treatment applied, those genes are not expressed and very likely there is no signal if we observe the tracks. On the other hand, the average tpm for the genes after applying the mc treatment, we can notice that there is some expression. To provide context, we observe the mean TPM distribution:

```{r mean_tpm,warning=FALSE}
DE_TPM %>% 
  unnest() %>% 
  dplyr::select(cell,batch,gene_id,mc,none) %>% 
  gather(treatment,mean_tpm,-(1:3)) %>% 
  separate(gene_id,c("ensembl","symbol"),sep = "\\_") %>% 
  mutate(batch = factor(batch,levels = c("old","new")),
         treatment = factor(treatment,levels = c("none","mc"))) %>% 
  filter(treatment == "mc") %>% 
  ggplot(aes(mean_tpm,cell,fill = cell))+
  geom_joy(stat = "binline",bins = 50)+
  facet_grid( batch ~ .,space = "free_y",scales = "free_y")+
  scale_x_continuous(limits = c(-5,3e2))+
  scale_fill_tech("google")+
  geom_vline(xintercept = 100,colour = "red",linetype= 2)




```

# Filtering out genes

```{r diff_genes_filter,cache=TRUE,dependson="tpm_summary"}

DE_results = inner_join(DE_results,
                        DE_TPM,by = c("cell","batch")) %>% 
  mutate(
    filtered_results = map(tpm.rank, filter, mc_perc >= none_perc - thr &
                         mc_perc <= none_perc + thr) %>% 
      map(dplyr::select,gene_id) %>% 
      map2(results,left_join,by = "gene_id") %>% 
      map(mutate,padj = p.adjust(pvalue)) %>% 
      map(filter,padj <= fdr)
  )

```

We filtered out genes for the GSEA analysis, by considering only the genes inside the orange lines that are differentially expressed (defined as genes with adj. p.value $\leq `r fdr`$):

```{r genes_considered}

DE_results %>% 
  mutate(
    total_genes = map_int(results,nrow),
    diff_genes =map_int(filtered_results,nrow)
  ) %>% 
  dplyr::select(batch,cell,contains("genes")) %>% 
  kable(digits = 0)


```


# GSEA analysis

```{r hallmark, include=FALSE,message=FALSE,cache=TRUE}
gmt_files = vector("character",length = 2)
gmt_files[1] = here::here("data/Pathway") %>% 
  list.files(full.names = TRUE,recursive = TRUE) %>% 
  str_subset("v5.1") %>% 
  str_subset(".gmt") %>% 
  str_subset("h.all") %>% 
  set_names(
    {
      basename(.) %>% 
        str_replace(".gmt","") %>% 
        str_split("\\.") %>% 
        map_chr( ~ .[length(.)])
    }
  ) %>% 
  pluck("symbols")

gmt_files[2] = here::here("data/Pathway") %>% 
  list.files(full.names = TRUE,recursive = TRUE) %>% 
  str_subset("Intere") %>% 
  set_names(
    "mark"
  )

pathways = gmt_files %>%
  set_names(c("hallmark","curated")) %>% 
  map(read.gmt) %>% 
  map(as_tibble)

```

## Hallmark analysis

```{r gsea_analysis,cache=TRUE,dependson="diff_genes_filter",include=FALSE}
name <- function(.x){
  nms = pull(.x,symbol)
  pull(.x,stat) %>% set_names(nms)
  
}

DE_results = DE_results %>% 
  mutate(
    gene_list = map(filtered_results, ~ arrange(.,desc(stat))) %>% 
      map(separate,gene_id,into = c("ensembl","symbol"),sep ="\\_") %>% 
      map(name)
  ) %>% 
  mutate(
    gsea_hallmark = map(gene_list,GSEA,TERM2GENE = pathways[["hallmark"]],
               verbose = FALSE),
    gsea_curated = map(gene_list,GSEA,TERM2GENE = pathways[["curated"]],
                       verbose = FALSE)) %>% 
  mutate_at(vars(contains("gsea")),
            funs( res = map(. , "result") %>% 
                    map(as_tibble)))

```



```{r gsea_hallmark_plots,include=TRUE,fig.width=9,fig.height=10}

plot_data = DE_results %>%
  dplyr::select(cell,batch,gsea_hallmark_res) %>% 
  mutate_if(is.list,
            funs( map(.,arrange,NES))
  ) %>%
  unnest() %>% 
  mutate(
    order = row_number()
  ) %>% 
  mutate(
    batch = factor(batch,levels = c("old","new"))
  )

plot_data %>% 
  ggplot(
    aes(order,NES,fill = pvalue)
  )+
  geom_bar(stat = "identity")+
  facet_grid( batch + cell ~ .,scales = "free",space = "free")+
  scale_fill_gradientn(colours = pal,
                        guide = guide_colourbar(title = NULL,
                        barheight = unit(0.8,"npc"),
                        barwidth = unit(0.01,"npc")))+
  coord_flip()+
  scale_x_continuous(
    expand = c(0,0),
    breaks = plot_data$order,
    labels = plot_data$ID
  )+
  theme_minimal()+
  theme(legend.position = "right",
      axis.title.y = element_blank())



```

## Curated analysis

```{r gsea_curated_plots,include=TRUE,fig.width=9}

plot_data = DE_results %>%
  dplyr::select(cell,batch,gsea_curated_res) %>% 
  mutate_if(is.list,
            funs( map(.,arrange,NES))
  ) %>%
  unnest() %>% 
  mutate(
    order = row_number()
  ) %>% 
  mutate(
    batch = factor(batch,levels = c("old","new"))
  )

plot_data %>% 
  ggplot(
    aes(order,NES,fill = pvalue)
  )+
  geom_bar(stat = "identity")+
  facet_grid( batch + cell ~ .,scales = "free",space = "free")+
  scale_fill_gradientn(colours = pal,
                        guide = guide_colourbar(title = NULL,
                        barheight = unit(0.8,"npc"),
                        barwidth = unit(0.01,"npc")))+
  coord_flip()+
  scale_x_continuous(
    expand = c(0,0),
    breaks = plot_data$order,
    labels = plot_data$ID
  )+
  theme_minimal()+
  theme(legend.position = "right",
      axis.title.y = element_blank())



```


# Further analysis

```{r log2fc_pathway,include=FALSE}

pathways = DE_results %>% 
  dplyr::select(batch,cell,ends_with("res")) %>% 
  gather(  gene_set,results,-(1:2)) %>% 
  mutate(
    gene_set = gene_set %>% str_replace("gsea_","") %>% 
      str_replace("_res","")
  ) %>% 
  unnest()%>% 
  dplyr::select(cell,batch,gene_set,ID,core_enrichment) %>% 
  mutate(
    core_enrichment = map(core_enrichment, str_split,"/") %>% 
      map(unlist)
  ) %>% 
  nest(-ID,-gene_set) %>% 
  mutate(
    core_enrichment = map(data,pull,core_enrichment) %>% 
      map(unlist) %>% 
      map(unique) %>% 
      map( ~ tibble(symbol = .))
  ) 

fold_change = DE_results %>% 
  dplyr::select(cell,batch,results) %>% 
  unnest() %>% 
  separate(gene_id,c("ensembl","symbol"),sep = "\\_") %>% 
  dplyr::select(cell,batch,symbol,log2FoldChange) %>% 
  mutate(
    batch = factor(batch,levels = c("old","new"))
  ) %>% 
  dplyr::rename(log2FC = log2FoldChange)

pathways = pathways %>% 
  dplyr::rename(fc = core_enrichment) %>% 
  mutate(
    fc = map(fc,inner_join,fold_change,by = "symbol"),
    new_batch_summary = map(fc,filter,batch == "new") %>% 
      map(dplyr::select,-batch) %>% 
      map(spread,cell,log2FC),
    EBV = map_dbl(new_batch_summary,
                       ~ norm(as.matrix(pull(.,dRdZ) - pull(.,EBV)  ))),
    NOKS = map_dbl(new_batch_summary,
                        ~ norm(as.matrix(pull(.,dRdZ) - pull(.,NOKS)))),
    dRdZ = 0
  )


```

## Hallmark pathways

```{r include=TRUE,include=TRUE,fig.width=7}

pathways %>% 
  filter(gene_set == "hallmark") %>% 
  dplyr::select(ID,EBV,NOKS,dRdZ) %>% 
  gather(cell,diff,-ID) %>% 
  ggplot( aes(reorder(ID,diff),diff,fill = cell))+
  geom_col(position = "dodge")+
  coord_flip()+
  xlab("ID")+
  theme(
    axis.title.x = element_blank()
  )+
  scale_fill_tech("google")

```



```{r include=TRUE,fig.height=6,fig.width=12,echo=FALSE,warning=FALSE,message=FALSE,results='hide'}

pathways = pathways %>% 
  mutate(
    plots = map2(fc,ID, 
                 ~ .x %>% 
                   ggplot(aes(reorder(symbol,-log2FC),
                              log2FC,colour = cell,shape = batch))+
                   geom_point(size = 2.5)+
                   theme(
                     axis.text.x = element_text(angle = 90,
                                                hjust = 1,vjust = .5 ) )+
                   scale_color_tech("google")+
                   scale_shape_manual(values = c(5,1))+
                   facet_grid(batch ~ . )+
                   xlab("gene")+
                   geom_hline(yintercept = 0,linetype = 2)+
                   ggtitle(.y)))

pathways %>% 
  filter(gene_set == "hallmark") %>% 
  pull(plots)

```

## Curated pathways

```{r include=TRUE,include=TRUE,fig.width=7}

pathways %>% 
  filter(gene_set != "hallmark") %>% 
  dplyr::select(ID,EBV,NOKS,dRdZ) %>% 
  gather(cell,diff,-ID) %>% 
  ggplot( aes(reorder(ID,diff),diff,fill = cell))+
  geom_col(position = "dodge")+
  coord_flip()+
  xlab("ID")+
  theme(
    axis.title.x = element_blank()
  )+
  scale_fill_tech("google")

```

```{r include=TRUE,fig.height=6,fig.width=12,echo=FALSE,warning=FALSE,message=FALSE,results='hide'}

pathways %>% 
  filter(gene_set != "hallmark") %>% 
  pull(plots)

```

