---
title: "RNAseq TPM exploratory"
output: html_notebook
---

```{r start,echo = FALSE}
  library(knitr)
  library(tidyverse)
  library(corrplot)

  opts_chunk$set(include = TRUE,echo = FALSE,eval = TRUE)
  theme_set(theme_minimal())
  v = viridisLite::viridis(50)[1]

```

# TPM distribution per dataset

We summarized RSEM's outcome into a matrix, where the rows are the genes and the columns are RNA-seq samples. First, we explore the quartiles of the different TPM distributions.

```{r libraries}

  mat = read_tsv("../data/metadata/RSEM_gene_TPM_matrix.tsv") 
  mat = mat %>% rename(gene = gene_id , transcript = `transcript_id(s)`)

  tpm = mat %>% gather(variable ,tpm , -c(1:2))
  
  distr = tpm %>% split(.$variable) %>% map(.f = function(.)summary(.$tpm) %>% round(2)) %>% bind_rows %>%
    mutate(summary = c("min","Q25","Q50","mean","Q75","max")) %>% select(summary,everything())
  distr

```

```{r tpm_analysis,fig.height=30,fig.width=10,fig.cap="TPM distribution per sample"}

  tpm %>% 
  ggplot(aes(tpm))+
    geom_histogram(bins = 50,fill = "white",colour = v)+
    xlim(0,1e2)+ylim(0,2e3)+
    facet_wrap(~variable,scales = "free_y",ncol = 3)
```

# Correlation matrix


```{r corplot,fig.width=12,fig.height=12}

M = mat %>% select(-gene,-transcript) %>% cor
corrplot(M,col = rev(viridisLite::viridis(30)),tl.col = v,
         tl.cex = .8,type = "upper",cl.lim = c(0,1), method = "ellipse",is.corr = FALSE,
         order = "hclust")

```


















