---
title: "Comparison by treatment"
output:
  html_document:
    theme: simplex
    toc: yes
  pdf_document:
    toc: yes
---

# Intro

I guess the idea is to find genes such that show high expression under the treatment, but at the same time exhibit low expression without the treatment, i.e. we are going to focus on the genes that are on the Wald's statistic distribution's tails. For that purpose, we are going to:

1. Perform contrast of MC-treated vs untreated samples for each cell line.

2. Compare the Wald's t-statistic between the three cell lines.


```{r load,include=FALSE,echo=FALSE,eval=TRUE}
library(magrittr)
library(tidyverse)
library(ghibli)
library(here)
library(viridis)
library(ggtech)
library(hexbin)
library(rlang)
library(rwlib)
library(clusterProfiler)
library(grid)
library(gridExtra)
library(scales)
library(knitr)
library(DESeq2)

pal = viridis(1e3, option = "D")

theme_set(theme_minimal() + 
            theme(legend.position = "top",
                  axis.text = element_text(size = 14),
                  axis.title = element_text(size = 16),
                  legend.title = element_text(size = 16),
                  legend.text = element_text(size = 14),
                  strip.text = element_text(size = 16)))

knitr::opts_chunk$set(eval = TRUE,
                      echo = FALSE,
                    include = TRUE,
                    warning = FALSE,
                    message = FALSE)

thr = 20
```

```{r diff_expression_analyis, cache=TRUE, include=FALSE}
rsem_dr = here("data/RSEM/hg19/Sept17")
align_dr = here("manuscript/logs/Sept17")

alignment = align_dr %>% 
  list.files(full.names = TRUE) %>% 
  stri_subset(".sh") %>% 
  tibble(
    file = .
  ) %>% 
  mutate(
    stats = map(file,read_tsv,col_names = FALSE,progress = FALSE))

parse_stats = function(stats)
{
  stats %>% 
    filter( 
      str_detect(X1,"total") | str_detect(X1,"mapped")) %>% 
    filter(
      negate(str_detect)(X1,"mate")
    ) %>% 
    mutate(
      type = if_else(str_detect(X1,"total"),"total","aligned"),
      reads = str_split(X1," ") %>% 
        map_chr( ~ .[1]) %>% 
        as.numeric()) %>% 
    dplyr::select(-X1) 
}

alignment %<>% 
  mutate(
    stats = map(stats,parse_stats),
    file = basename(file)
  ) %>% 
  unnest() %>% 
  spread(type,reads) %>% 
  mutate(
    perc = aligned / total,
    file = str_replace(file,".logs","")
  ) %>% 
  mutate(
    treatment = file %>% 
      str_detect("methyl") %>% 
      if_else(
        "methyl","none"
      ),
    cell = case_when(
          str_detect(file,"clone") ~ "dRdZ",
          str_detect(file,"akata") ~ "EBV" ,
          TRUE ~ "NOKS"),
    rep = file %>% 
      str_split("-") %>% 
      map_chr( ~ .[length(.)]) %>% 
      str_replace("clone","") %>% 
      str_replace("rep","") %>% 
      as.integer())


rsem_data = tibble(
  file = list.files(rsem_dr,full.names = TRUE)) %>% 
  mutate(
    rsem = map(file,read_tsv)
  ) %>% 
  mutate(
    file = basename(file) %>% 
      str_replace(".genes.results","")
  ) %>% 
  inner_join(
    select_if(alignment,negate(is.list)), by ="file"
  ) %>% 
  dplyr::select(-aligned,-total,-perc)

## remove contaminated samples clone2 and clone4
alignment %<>% 
  filter(!( cell == "dRdZ" & rep %in% c(2,4) ))

rsem_data %<>% 
  filter(!( cell == "dRdZ" & rep %in% c(2,4) ))


as_matrix <- function(x)
{
  x %>% 
    as.data.frame() %>% 
    tibble::remove_rownames() %>% 
    tibble::column_to_rownames("gene_id") %>% 
    as.matrix()
}

count_matrix = rsem_data %>% 
  dplyr::select(file,rsem) %>% 
  unnest() %>% 
  dplyr::select(file,gene_id,expected_count) %>% 
  mutate(
    expected_count = floor(expected_count)
  ) %>% 
  spread(file,expected_count) %>% 
  as_matrix()

coldata = rsem_data %>% 
  dplyr::select(file,cell,treatment) %>% 
  mutate(interac = paste(cell,treatment, sep = ".")) %>% 
  as.data.frame() %>% 
  tibble::remove_rownames() %>% 
  tibble::column_to_rownames("file")


## remove genes with low read counts, filtered genes with no reads but probably can 
## filter genes with low expression

deseq = DESeqDataSetFromMatrix(
  count_matrix,colData = coldata,
  design = ~ interac
)

deseq = deseq[ rowSums(assay(deseq) ) > thr,]
deseq = DESeq(deseq)


do_contrast = function(deseq,contrast)
{
  results(deseq,
          cooksCutoff = FALSE,
          contrast = contrast,
          tidy = TRUE) %>% 
    as_tibble() %>% 
    dplyr::rename(
      gene_id = row
    ) %>% 
    mutate(
      log10pval = -log10(pvalue)
    )
  
}

## The contrast is EBV.none vs NOKS.none

diff_genes = tribble(
  ~ cell, ~ test, ~ contrast,
  "dRdZ", "MC_vs_none", c("interac","dRdZ.methyl","dRdZ.none"),
  "EBV", "MC_vs_none",c("interac","EBV.methyl","EBV.none"),
  "NOKS","MC_vs_none",c("interac","NOKS.methyl","NOKS.none")
) %>% 
  mutate(
    results = map(contrast, ~ do_contrast(deseq,.))
  )

common_genes = diff_genes %>% 
  dplyr::select(results) %>% 
  unnest() %>% 
  bind_rows() %>% 
  group_by(gene_id) %>% 
  summarize(
    n = n()
  ) %>% 
  filter(n == 3) %>% 
  ungroup() %>% 
  dplyr::select(gene_id)

rsem_data = rsem_data %>% 
  mutate(
    rsem = map(rsem , inner_join,common_genes,by = "gene_id")
  )

rlog = rlog(deseq)  

rlogmat = rlog %>% 
  assay() %>% 
  as.matrix()

rm(coldata,common_genes,count_matrix,rlog)
```


```{r design, include=TRUE,fig.align="center",fig.width=5}

demo = alignment %>% 
  dplyr::select(cell,treatment,rep) %>% 
  group_by(cell,treatment) %>% 
  summarize( n = n()) %>% 
  ungroup()

demo %>% ggplot(aes( x = 1, y = 1,fill = interaction(cell,treatment)))+
  geom_rect(aes(xmin = .5,xmax = 1.5,ymin = .5,ymax = 1.5  ))+
  geom_text(aes(label = paste("Reps:",n,sep = "\n"),size = 40  ))+
  coord_fixed()+
  facet_grid(treatment ~ cell)+
  theme_void()+
  theme(legend.position = "none",
        strip.text = element_text(size = 25))+
  scale_fill_manual(values = c(ghibli_palette("PonyoMedium",6)[4:6],
                               ghibli_palette("PonyoLight",6)[4:6]))

```

# Wald's statistic analysis

The figure below shows the densities of the signal-to-noise meaures of the treated vs untreated contrast for all three cell lines. Clearly, all three cell lines resemble a similar pattern, with a slightly heavier tails for the NOKS case.

```{r walds,fig.align='center',fig.width=6,fig.height=5}

diff_genes %>% 
  dplyr::select(cell,results) %>% 
  unnest() %>% 
  ggplot(
    aes(stat,fill = cell)
  )+
  stat_density(geom = "polygon")+
  scale_fill_manual(values = ghibli_palette("PonyoLight",6)[4:6])+
  stat_density(geom = "line",colour = "black")+
  facet_grid( cell ~ . )+
  geom_vline(xintercept = 0,linetype = 2)+
  theme(
    strip.text.y = element_text(angle = 0)
  )

```

We then compare those three vectors, and we can notice that for most of the genes the signal-to-noise (i.e the log2FoldChange) share the same sign, which means that the MC treatment is affecting the majority of the genes in a similar fashion. This figure shows that the most differentially expressed genes are shared across cell lines, hence it is unlikely to find genes that are differentially expressed by the treatment in one cell line but not the other.


```{r scatter,fig.align='center',fig.width=8,fig.height=6.5,warning=FALSE}

diff_genes %>% 
  dplyr::select(cell, results) %>% 
  unnest() %>% 
  dplyr::select(cell,gene_id,stat) %>% 
  spread(cell,stat) %>% 
  dplyr::select(-gene_id) -> ss

scatters = list()
scatters[["NOKS_vs_dRdZ"]] = ss %>% ggplot(aes(NOKS,dRdZ))
scatters[["EBV_vs_dRdZ"]] = ss %>% ggplot(aes(EBV,dRdZ))
scatters[["EBV_vs_NOKS"]] = ss %>% ggplot(aes(EBV,NOKS))


pal = viridis(1000)

scatters %<>% map( ~ . + stat_binhex(bins = 51)+
                     coord_fixed()+
                     scale_x_continuous(limits = c(-22,22))+
                     scale_y_continuous(limits = c(-22,22))+
                     geom_vline(xintercept = 0,linetype = 2)+
                     geom_hline(yintercept = 0,linetype = 2)+
                     theme(legend.position = "right")+
                     scale_fill_gradientn(
                       colours = pal,trans = "log10",
                       labels =                               trans_format('log10',math_format(10^.x)),
                       guide = guide_colourbar(title = NULL,
                                               barheight = unit(0.3,"npc"),
                                               barwidth = unit(0.01,"npc"))))


corpl = ss %>% 
  cor() %>% 
  as.data.frame() %>% 
  rownames_to_column("cell1") %>% 
  gather(cell2,cor,-1) %>% 
  ggplot(aes(cell1,cell2,fill = cor))+
  geom_tile()+
  geom_text(aes(label = round(cor,2)),size = 8,
            colour = "grey")+
  theme(
    axis.title = element_blank(),
    legend.position = "right")+
  scale_fill_distiller(
    type = "div",
    palette = "RdBu",
    values = seq(0,1,length.out = 31),
    guide = guide_colourbar(title = NULL,
                            barheight = unit(0.3,"npc"),
                            barwidth = unit(0.01,"npc")))+
  coord_fixed()

grid.arrange(scatters[[1]],
             scatters[[2]],
             corpl,
             scatters[[3]],nrow = 2)

```


