---
title: "Comparison by treatment"
output:
  html_document:
    theme: simplex
    toc: yes
  pdf_document:
    toc: yes
---

# Intro

I guess the idea is to find genes such that show high expression under the treatment, but at the same time exhibit low expression without the treatment, i.e. we are going to focus on the genes that are on the Wald's statistic distribution's tails. For that purpose, we are going to:

1. Perform contrast of MC-treated vs untreated samples for each cell line.

2. Compare the Wald's t-statistic between the three cell lines.


```{r load,include=FALSE,echo=FALSE,eval=TRUE}
library(magrittr)
library(tidyverse)
library(ghibli)
library(here)
library(viridis)
library(ggtech)
library(hexbin)
library(rlang)
library(rwlib)
library(clusterProfiler)
library(grid)
library(gridExtra)
library(scales)
library(knitr)
library(DESeq2)

pal = viridis(1e3, option = "D")

theme_set(theme_minimal() + 
            theme(legend.position = "top",
                  axis.text = element_text(size = 14),
                  axis.title = element_text(size = 16),
                  legend.title = element_text(size = 16),
                  legend.text = element_text(size = 14),
                  strip.text = element_text(size = 16)))

knitr::opts_chunk$set(eval = TRUE,
                      echo = FALSE,
                    include = TRUE,
                    warning = FALSE,
                    message = FALSE)

thr = 20
```

```{r diff_expression_analyis, cache=TRUE, include=FALSE}
rsem_dr = here("data/RSEM/hg19/Sept17")
align_dr = here("manuscript/logs/Sept17")

alignment = align_dr %>% 
  list.files(full.names = TRUE) %>% 
  stri_subset(".sh") %>% 
  tibble(
    file = .
  ) %>% 
  mutate(
    stats = map(file,read_tsv,col_names = FALSE,progress = FALSE))

parse_stats = function(stats)
{
  stats %>% 
    filter( 
      str_detect(X1,"total") | str_detect(X1,"mapped")) %>% 
    filter(
      negate(str_detect)(X1,"mate")
    ) %>% 
    mutate(
      type = if_else(str_detect(X1,"total"),"total","aligned"),
      reads = str_split(X1," ") %>% 
        map_chr( ~ .[1]) %>% 
        as.numeric()) %>% 
    dplyr::select(-X1) 
}

alignment %<>% 
  mutate(
    stats = map(stats,parse_stats),
    file = basename(file)
  ) %>% 
  unnest() %>% 
  spread(type,reads) %>% 
  mutate(
    perc = aligned / total,
    file = str_replace(file,".logs","")
  ) %>% 
  mutate(
    treatment = file %>% 
      str_detect("methyl") %>% 
      if_else(
        "methyl","none"
      ),
    cell = case_when(
          str_detect(file,"clone") ~ "dRdZ",
          str_detect(file,"akata") ~ "EBV" ,
          TRUE ~ "NOKS"),
    rep = file %>% 
      str_split("-") %>% 
      map_chr( ~ .[length(.)]) %>% 
      str_replace("clone","") %>% 
      str_replace("rep","") %>% 
      as.integer())


rsem_data = tibble(
  file = list.files(rsem_dr,full.names = TRUE)) %>% 
  mutate(
    rsem = map(file,read_tsv)
  ) %>% 
  mutate(
    file = basename(file) %>% 
      str_replace(".genes.results","")
  ) %>% 
  inner_join(
    select_if(alignment,negate(is.list)), by ="file"
  ) %>% 
  dplyr::select(-aligned,-total,-perc)

## remove contaminated samples clone2 and clone4
alignment %<>% 
  filter(!( cell == "dRdZ" & rep %in% c(2,4) ))

rsem_data %<>% 
  filter(!( cell == "dRdZ" & rep %in% c(2,4) ))


as_matrix <- function(x)
{
  x %>% 
    as.data.frame() %>% 
    tibble::remove_rownames() %>% 
    tibble::column_to_rownames("gene_id") %>% 
    as.matrix()
}

count_matrix = rsem_data %>% 
  dplyr::select(file,rsem) %>% 
  unnest() %>% 
  dplyr::select(file,gene_id,expected_count) %>% 
  mutate(
    expected_count = floor(expected_count)
  ) %>% 
  spread(file,expected_count) %>% 
  as_matrix()

coldata = rsem_data %>% 
  dplyr::select(file,cell,treatment) %>% 
  mutate(interac = paste(cell,treatment, sep = ".")) %>% 
  as.data.frame() %>% 
  tibble::remove_rownames() %>% 
  tibble::column_to_rownames("file")


## remove genes with low read counts, filtered genes with no reads but probably can 
## filter genes with low expression

deseq = DESeqDataSetFromMatrix(
  count_matrix,colData = coldata,
  design = ~ interac
)

deseq = deseq[ rowSums(assay(deseq) ) > thr,]
deseq = DESeq(deseq)


do_contrast = function(deseq,contrast)
{
  results(deseq,
          cooksCutoff = FALSE,
          contrast = contrast,
          tidy = TRUE) %>% 
    as_tibble() %>% 
    dplyr::rename(
      gene_id = row
    ) %>% 
    mutate(
      log10pval = -log10(pvalue)
    )
  
}

## The contrast is EBV.none vs NOKS.none

diff_genes = tribble(
  ~ cell, ~ test, ~ contrast,
  "dRdZ", "MC_vs_none", c("interac","dRdZ.methyl","dRdZ.none"),
  "EBV", "MC_vs_none",c("interac","EBV.methyl","EBV.none"),
  "NOKS","MC_vs_none",c("interac","NOKS.methyl","NOKS.none")
) %>% 
  mutate(
    results = map(contrast, ~ do_contrast(deseq,.))
  )

common_genes = diff_genes %>% 
  dplyr::select(results) %>% 
  unnest() %>% 
  bind_rows() %>% 
  group_by(gene_id) %>% 
  summarize(
    n = n()
  ) %>% 
  filter(n == 3) %>% 
  ungroup() %>% 
  dplyr::select(gene_id)

rsem_data = rsem_data %>% 
  mutate(
    rsem = map(rsem , inner_join,common_genes,by = "gene_id")
  )

rlog = rlog(deseq)  

rlogmat = rlog %>% 
  assay() %>% 
  as.matrix()

rm(coldata,common_genes,count_matrix,rlog)
```


```{r design, include=TRUE,fig.align="center",fig.width=5}

demo = alignment %>% 
  dplyr::select(cell,treatment,rep) %>% 
  group_by(cell,treatment) %>% 
  summarize( n = n()) %>% 
  ungroup()

demo %>% ggplot(aes( x = 1, y = 1,fill = interaction(cell,treatment)))+
  geom_rect(aes(xmin = .5,xmax = 1.5,ymin = .5,ymax = 1.5  ))+
  geom_text(aes(label = paste("Reps:",n,sep = "\n"),size = 40  ))+
  coord_fixed()+
  facet_grid(treatment ~ cell)+
  theme_void()+
  theme(legend.position = "none",
        strip.text = element_text(size = 25))+
  scale_fill_manual(values = c(ghibli_palette("PonyoMedium",6)[4:6],
                               ghibli_palette("PonyoLight",6)[4:6]))

```

# Signal to noise analysis

The figure below shows the densities of the signal-to-noise measures of the treated vs untreated contrast for all three cell lines. Clearly, all three cell lines resemble a similar pattern, with a slightly heavier tails for the NOKS case.

```{r walds,fig.align='center',fig.width=6,fig.height=5}

diff_genes %>% 
  dplyr::select(cell,results) %>% 
  unnest() %>% 
  ggplot(
    aes(stat,fill = cell)
  )+
  stat_density(geom = "polygon")+
  scale_fill_manual(values = ghibli_palette("PonyoLight",6)[4:6])+
  stat_density(geom = "line",colour = "black")+
  facet_grid( cell ~ . )+
  geom_vline(xintercept = 0,linetype = 2)+
  theme(
    strip.text.y = element_text(angle = 0)
  )+
  xlab("Signal-to-noise")

```

We then compare those three vectors, and we can notice that for most of the genes the signal-to-noise (i.e the log2FoldChange) share the same sign, which means that the MC treatment is affecting the majority of the genes in a similar fashion. This figure shows that the most differentially expressed genes are shared across cell lines, hence it is unlikely to find genes that are differentially expressed by the treatment in one cell line but not the other.


```{r scatter,fig.align='center',fig.width=8,fig.height=6.5,warning=FALSE}

diff_genes %>% 
  dplyr::select(cell, results) %>% 
  unnest() %>% 
  dplyr::select(cell,gene_id,stat) %>% 
  spread(cell,stat) %>% 
  dplyr::select(-gene_id) -> ss

scatters = list()
scatters[["NOKS_vs_dRdZ"]] = ss %>% ggplot(aes(NOKS,dRdZ))
scatters[["EBV_vs_dRdZ"]] = ss %>% ggplot(aes(EBV,dRdZ))
scatters[["EBV_vs_NOKS"]] = ss %>% ggplot(aes(EBV,NOKS))


pal = viridis(1000)

scatters %<>% map( ~ . + stat_binhex(bins = 51)+
                     coord_fixed()+
                     scale_x_continuous(limits = c(-22,22))+
                     scale_y_continuous(limits = c(-22,22))+
                     geom_abline(slope = 1, intercept = 0,linetype = 2,
                                 colour = "orange",size = 1.2)+
                     geom_vline(xintercept = 0,linetype = 2)+
                     geom_hline(yintercept = 0,linetype = 2)+
                     theme(legend.position = "right")+
                     scale_fill_gradientn(
                       colours = pal,trans = "log10",
                       labels =                               trans_format('log10',math_format(10^.x)),
                       guide = guide_colourbar(title = NULL,
                                               barheight = unit(0.3,"npc"),
                                               barwidth = unit(0.01,"npc"))))


corpl = ss %>% 
  cor() %>% 
  as.data.frame() %>% 
  rownames_to_column("cell1") %>% 
  gather(cell2,cor,-1) %>% 
  ggplot(aes(cell1,cell2,fill = cor))+
  geom_tile()+
  geom_text(aes(label = round(cor,2)),size = 8,
            colour = "grey")+
  theme(
    axis.title = element_blank(),
    legend.position = "right")+
  scale_fill_distiller(
    type = "div",
    palette = "RdBu",
    values = seq(0,1,length.out = 31),
    guide = guide_colourbar(title = NULL,
                            barheight = unit(0.3,"npc"),
                            barwidth = unit(0.01,"npc")))+
  coord_fixed()

grid.arrange(scatters[[1]],
             scatters[[2]],
             corpl,
             scatters[[3]],nrow = 2)

```

The figure above exhibits that in the case of NOKS vs dRdZ, the signal to noise metrics are relatively symmetric around the diagonal, but in the case of the other two pairs (EBV vs NOKS, and EBV vs dRdZ) we can notice a slightly tilted distribution towards NOKS, and dRdZ respectively. Searching in the Bioconductor forums, I found the [following answer](https://support.bioconductor.org/p/61509/) by the creator of DESeq2, which states a framework to test ratio of ratio (which appears to be a common problem in RIP-seq / CLIP-seq), where the IP counts are normalized by the Input counts (very similar to ChIP-seq).

For that purpose, we fit a different models for each pair of cell lines, which means that we are testing the `cell:treatment` effect, which indicates that we are testing the `treatment` effect across `cell` lines.

```{r fun ,include=TRUE,echo=TRUE,eval=TRUE}
ratio_of_ratios_deseq <- function(rsem_data,thr = 20)
{
  count_matrix =  rsem_data %>% 
    dplyr::select(file,rsem) %>% 
    unnest() %>% 
    dplyr::select(file,gene_id,expected_count) %>% 
    mutate(
      expected_count = floor(expected_count)
      ) %>% 
    spread(file,expected_count) %>% 
    as_matrix()

  coldata = rsem_data %>% 
    dplyr::select(file,cell,treatment) %>% 
    mutate(interac = paste(cell,treatment, sep = ".")) %>% 
    as.data.frame() %>% 
    tibble::remove_rownames() %>% 
    tibble::column_to_rownames("file")

  deseq = DESeqDataSetFromMatrix(
    count_matrix,colData = coldata,
    design = ~ cell + treatment + cell:treatment)

  deseq = deseq[ rowSums(assay(deseq) ) > thr,]
  deseq = DESeq(deseq, test = "LRT", reduced =  ~ cell + treatment)

  deseq
  
}
```


```{r ratio_of_ratio_analysis,include=FALSE,echo=FALSE,eval=TRUE,cache=TRUE,dependson="diff_expression_analysis"}

ratio_of_ratios = tribble(
  ~ test , ~ deseq,
  "EBV_vs_NOKS" , 
  rsem_data %>% 
    filter(cell != "dRdZ") %>% 
    ratio_of_ratios_deseq(thr = thr),
  "EBV_vs_dRdZ" , 
  rsem_data %>% 
    filter(cell != "NOKS") %>% 
    ratio_of_ratios_deseq(thr = thr),
  "NOKS_vs_dRdZ",
  rsem_data %>% 
    filter(cell != "EBV") %>% 
    ratio_of_ratios_deseq(thr = thr)
)

clean_results <- function(deseq)
{
  results(deseq,
          cooksCutoff = FALSE,
          tidy = TRUE) %>% 
    as_tibble() %>% 
    dplyr::rename(
      gene_id = row
    ) %>% 
    mutate(
      log10pval = -log10(pvalue)
    )
  
}

ratio_of_ratios %<>% 
  mutate(
    results = map(deseq, clean_results) %>% 
      map(separate,gene_id , into = c("ensembl","symbol"),
          sep = "\\_",extra = "drop" )
  )

```

This test returns a smaller list of genes that are differentially expressed. For example, if we consider the genes that are differentially expressed with adjusted p.value $\leq 0.01$, we can notice that the number of genes that are differentially expressed is much larger when testing the MC effect across the EBV and NOKS cell lines, than when any of them is the mutant type dRdZ.

```{r venn,include=TRUE,fig.align='center',fig.width=4}

library(Vennerable)

ratio_of_ratios %>% 
  mutate(
    ll = map(results,filter,padj <= 1e-2) %>% 
      map(pull,symbol)
  ) %>% pull(ll) %>% 
  set_names(pull(ratio_of_ratios,test)) %>% Vennerable::Venn() -> venn_ratios

plot(venn_ratios)

```

Furthermore, we can examine the genes in these subgroups. For example, we can notice that the intersection of `EBV_vs_dRdZ` and `EBV_vs_NOKS` have `r venn_ratios@IntersectionSets[["110"]] %>% length()`, and in that group there are genes such that the distance between EBV and the other cell lines are maximized, thus it results in genes where the signal-to-noise is close between them (i.e. NOKS and dRdZ). Alternatively, in the regions of the Venn diagram where the genes are only differentiated in one category, we can observe that the cell line that was not considered is usually in the middle.

```{r,fig.width=9,fig.height=5,fig.align='center',results='hide'}

nms = c("100" = "EBV_vs_NOKS",
        "010" = "EBV_vs_dRdZ",
        "001" = "NOKS_vs_dRdZ",
        "110" = "EBV_vs_both",
        "101" = "NOKS_vs_both",
        "011" = "dRdZ_vs_both"
        )

venn_d_plot <- function(subset = c("100","010","001","011","101",
                                      "110","111"),
                           venn_ratios,diff_genes,   nms = NULL)
{
  subset = match.arg(subset)
  
  ss = venn_ratios@IntersectionSets[[subset]]
  
  nm = if_else(is.null(nms), subset,nms[subset])
  
  out = diff_genes %>% 
    mutate(results = map(results,separate,gene_id,into = c("ensembl","symbol"),
                         sep = "\\_",extra = "drop") %>% 
             map(filter,symbol %in% ss)) %>% 
    dplyr::select(cell,results) %>% 
    unnest() %>%
    mutate(symbol = fct_reorder(symbol,stat)) %>% 
    ggplot(aes(symbol , stat,colour = cell))+
    geom_point(size = 2)+
    theme(
      axis.text.x = element_text(angle = 90,vjust = .5,hjust= 1),
      axis.title.x = element_blank()
    )+
    scale_color_tech("google")+
    geom_hline(yintercept = 0,linetype = 2)+
    ylab("Signal-to-noise")+ggtitle(nm)
  
  if(length(ss) > 60){
    out = out + theme(axis.text.x = element_text(angle = 90,vjust = .5,hjust = 1,
                                                 size = 4))
  }
  out

}

venn_ratios@IntersectionSets %>% names() %>% 
{.[! . %in% c("000","111")]} %>% 
  map(  venn_d_plot,
        venn_ratios = venn_ratios,diff_genes = diff_genes,nms = nms)

```

```{r,fig.width=5,fig.height=9,include=FALSE,echo=FALSE,eval=FALSE}

## Deprecated code

library(ggrepel)

diff_genes %>% 
  dplyr::select(cell, results) %>% 
  unnest() %>% 
  dplyr::select(cell,gene_id,stat) %>% 
  spread(cell,stat) %>% 
  mutate(
    dRdZ_EBV = dRdZ / EBV - 1,
    dRdZ_NOKS = dRdZ / NOKS -1 ,
    NOKS_EBV = NOKS / EBV -1
  ) %>% 
  dplyr::select(-EBV,-NOKS,-dRdZ) %>% 
  mutate(id = seq_along(gene_id)) %>% 
  gather(test,value,-id,-gene_id) -> ss2

M = 100

gene_list = ss2 %>% 
  separate(gene_id,into = c("ensembl","gene_id"),sep = "\\_",extra = "drop") %>% 
  filter(!between(value,-M,M)) %>% 
  filter(abs(value) < 10000) %>% 
  pull(gene_id) %>% unique()

diff_genes %>% 
  dplyr::select(cell,results) %>% 
  unnest() %>% 
  separate(gene_id,into = c("ensem","gene_id"),sep = "\\_",extra = "drop") %>% 
  filter(gene_id %in% gene_list) %>% 
  filter(padj <= .1) %>% 
  dplyr::select(cell,gene_id,stat,padj) %>% 
  ggplot(aes(reorder(gene_id,-padj),-log10(padj)))+
  geom_col(aes(fill = cell),position = "dodge")+
  coord_flip()


```

# Pathway analysis

This analysis returns a different signal-to-noise metric for each cell line. Hence, we are capable of performing a pathway analysis too. For example, we can notice that the none of the `MYC_TARGETS_V1` or `MYC_TARGETS_V2` pathways are enriched in the `EBV_vs_dRdZ` test, but the first one is enriched in the `EBV_vs_NOKS` test and both are enriched in the `NOKS_vs_dRdZ` test.

```{r gsea, message=FALSE,include=FALSE,eval=TRUE,echo=FALSE}

gmt_files =  here::here("data/Pathway") %>% 
  list.files(full.names = TRUE,recursive = TRUE) %>% 
  stri_subset("kera") %>% 
  set_names(
    {
      c("hallmark","curated")
    }
  )

pathways = tibble(
 pathway_name = names(gmt_files), files = gmt_files
) %>% 
  mutate(
    pathways = files %>% 
      map(read.gmt) %>% map(as_tibble)
  ) %>% 
  dplyr::select(-files) 

name <- function(.x){
  nms = pull(.x,symbol)
  pull(.x,stat) %>% set_names(nms)
  
}

ratio_of_ratios %<>% 
  crossing(pathways) %>% 
  mutate(
    genes_list = map(results,arrange,desc(stat) ) %>% map(name),
    gsea = map2( genes_list, pathways,
                 ~ GSEA(.x,TERM2GENE = .y,verbose = FALSE)),
    gsea_results = map(gsea,"result") %>% map(as_tibble))

```



```{r pathway_NES,include=TRUE,echo=FALSE,fig.align="center",fig.width=8,fig.height=12}

gsea_data = ratio_of_ratios %>% 
  dplyr::filter(pathway_name == "hallmark") %>% 
  dplyr::select(test,gsea_results) %>% 
  mutate(
    gsea_results = map(gsea_results,arrange,NES)) %>%
  unnest() %>% 
  mutate(order = seq_along(NES))


gsea_data %>% 
  ggplot( aes(order,NES,fill = pvalue))+
  geom_bar(stat = "identity")+
  scale_fill_gradientn(colours = viridis::viridis(100),
                       guide = guide_colourbar(title = NULL,
                                                 barheight = unit(0.8,"npc"),
                        barwidth = unit(0.01,"npc")))+
  coord_flip()+
  scale_x_continuous(    expand = c(0,0),
                         breaks = gsea_data$order,
                         labels = gsea_data$ID)+
  facet_grid(test ~ . ,scales = "free_y",space = "free_y")+
  theme_minimal()+
  theme(legend.position = "right",
        axis.text.y = element_text(size = 10),
        axis.title.y = element_blank())
    


```


## GSEA plots for `MYC_TARGETS`:

```{r myc_pathways, include=TRUE,echo=FALSE,eval=TRUE,fig.width=10,fig.height=8,warning=FALSE}
runningScore_comparison = function(geneSetID, gsea_tibb)
{

  gsea_tibb %<>%
    mutate(score_plots = map(gsea, 
                             ~ gseaplot(., geneSetID,by = "runningScore")) %>%
             map2(test , ~ .x + ylab(.y)),
           prerank_plots = map2(gsea,test,
                                ~ gseaplot(.,geneSetID,by = "preranked")) %>% 
             map2(test, ~ .x + ylab(.y)))
  
  grid.arrange(gsea_tibb$score_plots[[1]],
               gsea_tibb$prerank_plots[[1]],
               gsea_tibb$score_plots[[2]],
               gsea_tibb$prerank_plots[[2]],
               gsea_tibb$score_plots[[3]],
               gsea_tibb$prerank_plots[[3]],top = geneSetID,
               nrow = 3)
}

hallmark_gsea = ratio_of_ratios %>% 
  filter(pathway_name == "hallmark") %>% 
  dplyr::select(test,gsea)

runningScore_comparison("HALLMARK_MYC_TARGETS_V1",hallmark_gsea)


```


```{r demo_myc2,include=TRUE,echo=FALSE,eval=TRUE,fig.width=10,fig.height=8,warning=FALSE}

runningScore_comparison("HALLMARK_MYC_TARGETS_V2",hallmark_gsea)


```

### GSEA plot for `DNA_REPAIR`

I picked this pathway because it appear in both of the treated vs untreated pathways of Scott's cell lines, but in none of your data.

```{r demo_bad,include=TRUE,echo=FALSE,eval=TRUE,fig.width=10,fig.height=8,warning=FALSE}

runningScore_comparison("HALLMARK_DNA_REPAIR",hallmark_gsea)


```


```{r bio_theme_comparison,fig.align='center',fig.height=9,fig.width=7.5}

process_results = function(gsea)
{

  res = gsea@result  %>% 
    as_data_frame()
  
  out = res

  if(nrow(out) > 0){
    out %<>% 
      mutate(
        core_enrichment = str_split(core_enrichment,"/"),
        ngenes = map_dbl(core_enrichment,length),
        ratio = ngenes / setSize
        )
  }
  
  out
  
}

gsea_pathways = ratio_of_ratios %>% 
  filter(pathway_name == "hallmark") %>% 
  mutate(
    clean_res = map(gsea,process_results)
  )



gsea_pathways %>% 
  filter(test != "EBV_vs_NOKS") %>% 
  dplyr::select(test,clean_res) %>% 
  unnest() %>% 
  ggplot(aes(test,reorder(ID,-p.adjust),size = ratio))+
  geom_point(aes(colour = p.adjust))+
  theme(
    legend.position=c(1,1), 
    legend.justification=c(0, 1), 
    legend.key.width=unit(1, "lines"), 
    plot.margin = unit(c(1, 5, 0.5, 0.5), "lines"),
    legend.box = "vertical",
    axis.title = element_blank(),
    axis.text.x = element_text(angle = 90,hjust = 1,vjust = .5)
  )+
  scale_color_distiller(palette = "Spectral",
                        guide = guide_colourbar(title = "padj",
                                                barheight = unit(0.5,"npc"),
                                                barwidth = unit(0.01,"npc")))

```


```{r bio_theme_comp_curated,fig.align='center',fig.width=7.5,fig.height=8}

gsea_pathways_curated = ratio_of_ratios %>% 
  filter(pathway_name != "hallmark") %>% 
  mutate(
    clean_res = map(gsea,process_results)
  )

gsea_pathways_curated %>% 
  filter(test != "EBV_vs_NOKS") %>% 
  dplyr::select(test,clean_res) %>% 
  unnest() %>% 
  ggplot(aes(test,reorder(ID,-p.adjust),size = ratio))+
  geom_point(aes(colour = p.adjust))+
  theme(
    legend.position=c(1,1), 
    legend.justification=c(0, 1), 
    legend.key.width=unit(1, "lines"), 
    plot.margin = unit(c(1, 5, 0.5, 0.5), "lines"),
    legend.box = "vertical",
    axis.title = element_blank(),
    axis.text.x = element_text(angle = 90,hjust = 1,vjust = .5)
  )+
  scale_color_distiller(palette = "Spectral",
                        guide = guide_colourbar(title = "padj",
                                                barheight = unit(0.5,"npc"),
                                                barwidth = unit(0.01,"npc")))


```



```{r venn_pathway, include=TRUE,fig.align='center',fig.width=4}

gsea_pathways %<>% 
  filter(test != "EBV_vs_NOKS") %>% 
  dplyr::select(test,clean_res) %>% 
  mutate(
    ID = map(clean_res,pull,ID)
  )

pull(gsea_pathways,ID) %>% 
  set_names(pull(gsea_pathways,test)) %>% 
  Vennerable::Venn() -> venn_ratios2

plot(venn_ratios2)




```

Brief note: There are `r venn_ratios2@IntersectionSets[["11"]] %>% length()` shared pathways, `r venn_ratios2@IntersectionSets[["10"]] %>% length()` that are EBV vs dRdZ specific, and `r venn_ratios2@IntersectionSets[["01"]] %>% length()` that are NOKS vs dRdZ specific. This suggests, that dRdZ is more similar than NOKS as there is a smaller amount of enriched pathways that turn on by the genes that show a significant interaction between dRdZ and NOKS when the treatment in applied.


```{r pathway_NES2,include=TRUE,echo=FALSE,fig.align="center",fig.width=8,fig.height=2}

gsea_data = ratio_of_ratios %>% 
  dplyr::filter(pathway_name != "hallmark") %>% 
  dplyr::select(test,gsea_results) %>% 
  mutate(
    gsea_results = map(gsea_results,arrange,NES)) %>%
  unnest() %>% 
  mutate(order = seq_along(NES))

gsea_data %>% 
  ggplot( aes(order,NES,fill = pvalue))+
  geom_bar(stat = "identity")+
  scale_fill_gradientn(colours = viridis::viridis(100),
                       guide = guide_colourbar(title = NULL,
                                                 barheight = unit(0.8,"npc"),
                        barwidth = unit(0.01,"npc")))+
  coord_flip()+
  scale_x_continuous(    expand = c(0,0),
                         breaks = gsea_data$order,
                         labels = gsea_data$ID)+
  facet_grid(test ~ . ,scales = "free_y",space = "free_y")+
  theme_minimal()+
  theme(legend.position = "right",
        axis.text.y = element_text(size = 10),
        strip.text.y = element_text(angle = 0),
        axis.title.y = element_blank())
    


```

# Demo gene coefficients

Lets recall the DESeq model, for gene $i$ and sample $j$:

$$
\begin{aligned}
K_{ij} &\sim \mbox{NB}(\mu_{ij}, \alpha_i) \\
\mu_{ij} &= s_j q_{ij} \\
\log_2(q_{ij}) &= \mathbf{x}_{j\cdot}\beta_i
\end{aligned}
$$
where:

- $K_{ij}$ are the counts of gene $i$ and sample $j$
- $\mu_{ij}$ is the fitted mean, and $\alpha_i$ is a gene-specific dispersion parameter

The dispersion specific parameters are obtained by fitting a model of the following relationship:

$$
\mbox{Var}(K_{ij}) = \mu_{ij} + \alpha_i \mu_{ij}^2
$$

From this description of the model, we can notice that:

1. For every gene, we estimate a different model defined by its mean and dispersion parameters. 

2. This model differs from a typical linear regression model in the estimation of the variance. From the previous equation, we can observe that the estimated variance of the counts for gene $i$ and sample $j$ depends on the estimated mean. Hence, when testing for the significance of more than one terms it is necessary to use different methods.

3. The method that we used (suggested in [here](https://support.bioconductor.org/p/61509/)) is the LRT. Which sort of works like an ANOVA test, as it always positive and we are comparing the likelihood of the full model `~ cell + treatment + cell:treatment` with the likelihood of the reduced model `~ cell + treatment`.

As an example, we are going to consider the test that we did for the ratios of EBV vs NOKS (respect to treatment), we can see that the size factors are a slight correction (i.e. close to 1) but we are still working in the count scale:

```{r deseq_demo, include=TRUE,echo=TRUE,eval=TRUE}

demo_deseq = ratio_of_ratios$deseq[[1]]  ## we already fit this one
colData(demo_deseq)

```

For every gene, we are testing the interaction effect between the reduced model `~ cell + treatment` and the full model `~ cell + treatment + cell:treatment`. The model matrices are:

- Reduced model `~ cell + treatment`

```{r , include=TRUE,eval=TRUE,echo=TRUE}

with(colData(demo_deseq), model.matrix( ~ cell + treatment))

```

- Full model `~ cell + treatment + cell:treatment`

```{r , include=TRUE,eval=TRUE,echo=TRUE}

with(colData(demo_deseq), model.matrix( ~ cell + treatment + cell:treatment))

```

For example, we can check for a few genes:
```{r countmatrix, include=TRUE,echo=TRUE,eval=TRUE}

count_matrix = assay(demo_deseq,"counts")

geneID = "MAP1B"
log2(count_matrix)[ rownames(count_matrix) %>% 
          str_detect(geneID),]

geneID = "MPZL3"
log2(count_matrix)[ rownames(count_matrix) %>% 
          str_detect(geneID),]


```


```{r coeff_matrix, include=TRUE,echo=TRUE,eval=TRUE}

coeffs = coef(demo_deseq)

geneID = "MAP1B"
coeffs[ rownames(coeffs) %>% 
          str_detect(geneID),]

geneID = "MPZL3"
coeffs[ rownames(coeffs) %>% 
          str_detect(geneID),]


```








