---
title: "Effect of EBV infection in dRdZ mutant"
output: 
  html_document:
    toc: yes
    theme: flatly
---

# Introduction

From Eric's email:

> What the figure aims to show is that the effect of EBV infection is lost with the dRdZ mutant which looks more like an uninfected NOK with respect to the change in the signature genes upon MC treatment.

We prepare the data, basically load and remove genes with very low counts, 

```{r load,include=FALSE,echo=FALSE,eval=TRUE}
library(tidyverse)
library(DESeq2)
library(here)
library(Vennerable)
library(gridExtra)
library(ggrepel)

theme_set(theme_bw())

knitr::opts_chunk$set(eval = TRUE,
                      echo = FALSE,
                    include = TRUE,
                    warning = FALSE,
                    message = FALSE)


```


```{r prepare_data,message=FALSE,warning=FALSE,cache =TRUE,include=FALSE}
rsem_dr = here("data/RSEM/hg19/Sept17")
align_dr = here("manuscript/logs/Sept17")

## alignment = read_tsv("figs/fig2/Scott_aligned_reads_table.tsv")

stri_subset = function(string,pattern)string[negate(str_detect)(string,pattern)]

alignment = align_dr %>% 
  list.files(full.names = TRUE) %>% 
  stri_subset(".sh") %>% 
  tibble(
    file = .
  ) %>% 
  mutate(
    stats = map(file,read_tsv,col_names = FALSE))

parse_stats = function(stats)
{
  stats %>% 
    filter( 
      str_detect(X1,"total") | str_detect(X1,"mapped")) %>% 
    filter(
      negate(str_detect)(X1,"mate")
    ) %>% 
    mutate(
      type = if_else(str_detect(X1,"total"),"total","aligned"),
      reads = str_split(X1," ") %>% 
        map_chr( ~ .[1]) %>% 
        as.numeric()) %>% 
    select(-X1) 
}

alignment = alignment %>% 
  mutate(
    stats = map(stats,parse_stats),
    file = basename(file)
  ) %>% 
  unnest() %>% 
  spread(type,reads) %>% 
  mutate(
    perc = aligned / total,
    file = str_replace(file,".logs","")
  ) %>% 
  mutate(
    treatment = file %>% 
      str_detect("methyl") %>% 
      if_else(
        "methyl","none"
      ),
    cell = case_when(
          str_detect(file,"clone") ~ "dRdZ",
          str_detect(file,"akata") ~ "EBV" ,
          TRUE ~ "NOKS"),
    rep = file %>% 
      str_split("-") %>% 
      map_chr( ~ .[length(.)]) %>% 
      str_replace("clone","") %>% 
      str_replace("rep","") %>% 
      as.integer())


rsem_data = tibble(
  file = list.files(rsem_dr,full.names = TRUE)) %>% 
  mutate(
    rsem = map(file,read_tsv,progress = FALSE)
  ) %>% 
  mutate(
    file = basename(file) %>% 
      str_replace(".genes.results","")
  ) %>% 
  inner_join(
    select_if(alignment,negate(is.list)), by ="file"
  ) %>% 
  select(-aligned,-total,-perc)

## remove contaminated samples clone2 and clone4
alignment = alignment %>% 
  filter(!( cell == "dRdZ" & rep %in% c(2,4) ))

rsem_data = rsem_data %>% 
  filter(!( cell == "dRdZ" & rep %in% c(2,4) ))

rm(align_dr,alignment)

as_matrix <- function(x)
{
  x %>% 
    as.data.frame() %>% 
    tibble::remove_rownames() %>% 
    tibble::column_to_rownames("gene_id") %>% 
    as.matrix()
}

count_matrix = rsem_data %>% 
  select(file,rsem) %>% 
  unnest() %>% 
  select(file,gene_id,expected_count) %>% 
  mutate(
    expected_count = floor(expected_count)
  ) %>% 
  spread(file,expected_count) %>% 
  as_matrix()

coldata = rsem_data %>% 
  select(file,cell,treatment) %>% 
  mutate(interac = paste(cell,treatment, sep = ".")) %>% 
  as.data.frame() %>% 
  tibble::remove_rownames() %>% 
  tibble::column_to_rownames("file") %>% 
  mutate_if(is_character,as.factor)

```

Since we want to obtain the genes that are differentially expressed (upregulated / downregulated) due to the treatment's effect, we need to prepare two models, one coded in terms of the full model and another in terms of the interactions. 

```{r deseq,cache = TRUE,dependson="prepare_data",message=FALSE}
deseq = DESeqDataSetFromMatrix(
  count_matrix,colData = coldata,
  design = ~ interac
)

thr = 20

deseq = deseq[ rowSums(assay(deseq) ) > thr,]
deseq_full = deseq
design(deseq_full) = ~ cell + treatment + cell * treatment

deseq = DESeq(deseq,quiet = TRUE)
deseq_full = DESeq(deseq_full,quiet = TRUE)
```


We are going to use the following function to test for hypothesis, contrast needs to be a vectors of the type `c("cat","A","B")`, and is going to test `A vs B` in the `cat` category.

```{r contrasts}

do_contrast = function(deseq,contrast)
{
  results(deseq,
          cooksCutoff = FALSE,
          contrast = contrast,
          tidy = TRUE) %>% 
    as_tibble() %>% 
    dplyr::rename(
      gene_id = row
    ) %>% 
    mutate(
      log10pval = -log10(pvalue)
    ) %>% 
    separate(gene_id,into = c("ensembl","symbol"),sep = "\\_",extra = "drop")
  
}

```

1. Get differentially expressed genes for all genes, and by cell

```{r treatment_genes}

all_diff_genes = do_contrast(deseq_full,c("treatment","methyl","none"))
ebv_diff_genes = do_contrast(deseq, c("interac","EBV.methyl","EBV.none"))
noks_diff_genes = do_contrast(deseq, c("interac","NOKS.methyl","NOKS.none"))
drdz_diff_genes = do_contrast(deseq, c("interac","dRdZ.methyl","dRdZ.none"))

```

    - `all_diff_genes` tests the hypothesis of the treatment effect is zero, regardless of treatment

    - `cell_diff_genes`  tests the hypothesis of the treatment effect is zero, when the cell is specified.

2. Compare EBV and NOKS differentially expressed genes vs all by `methyl` treat. 

```{r venn,out.height=2}

symbols_expressed = function(results, fdr, upregulated = c(NA,TRUE,FALSE))
{
  results = results %>% 
    filter(padj <= fdr)
  
  if(!is.na(upregulated)){
    if(upregulated)results = results %>% filter(log2FoldChange > 0)
    else results = results %>% filter(log2FoldChange < 0 )
  }
  
  results %>% 
    pull(symbol)
  
}

fdr = 0.05
upr = TRUE
Vupr_genes = list(
  "all" = symbols_expressed(all_diff_genes,fdr,upr),
  "ebv" = symbols_expressed(ebv_diff_genes,fdr,upr),
  "noks" = symbols_expressed(noks_diff_genes,fdr,upr),
  "drdz" = symbols_expressed(drdz_diff_genes,fdr,upr)
) %>% Venn()

upr = FALSE
Vdown_genes = list(
  "all" = symbols_expressed(all_diff_genes,fdr,upr),
  "ebv" = symbols_expressed(ebv_diff_genes,fdr,upr),
  "noks" = symbols_expressed(noks_diff_genes,fdr,upr),
  "drdz" = symbols_expressed(drdz_diff_genes,fdr,upr)
) %>% Venn() 



```

    - Upregulated genes


```{r upr_genes}


plot(Vupr_genes,doWeights = FALSE,
     show = list(Faces = FALSE))



```

    - Downregulated genes
  
```{r down_reg}
plot(Vdown_genes,doWeights = FALSE,
     show = list(Faces = FALSE))

```

```{r log2fun,include=FALSE}

diff_genes = tribble(
  ~ gene_set , ~ results,
  "all", all_diff_genes,
  "ebv", ebv_diff_genes,
  "noks", noks_diff_genes,
  "drdz", drdz_diff_genes
) %>% 
  unnest()


log2FC_plot <- function(gene_symbols,diff_genes, log = FALSE)
{
  dt = diff_genes %>% 
    filter(symbol %in% gene_symbols) %>% 
    filter(gene_set != "all") %>% 
    mutate(symbol = factor(symbol,levels = gene_symbols),
           cond = rep(log,nrow(.)),
           respo = if_else(cond,log2FoldChange,2^log2FoldChange)) 
  dt %>% 
    ggplot(aes(symbol,respo,colour = gene_set))+
    geom_point(size = 3, shape = 1)+
    scale_color_brewer(palette = "Set1",name = "cell")+
    theme(
      legend.position = "top",
      axis.text.x = element_text(angle = 30,hjust = 1)
    )+
    ylab(if_else(log,"log2 Fold Change","Fold Change"))
  
}


```

# Recovery: plot from slide    
  
```{r prev_plot,fig.height=4,fig.width=7,results='hold'}  

myc_pathway = c("MYC","MYBBP1A","MCM6","KPNA2",
                "KIF4A","HMGB2","CDK4")

log2FC_plot(myc_pathway,diff_genes)
log2FC_plot(myc_pathway,diff_genes,TRUE)


```

- The ranges and the order of the cells kinda agree, but using the model estimated fold changes, we can see that EBV is not at the top for every gene, where `KPNA2` could be the most notable example.

## Gene sets

We take the genes that are differentially expressed in EBV or NOKS but not the rest.

```{r venn_genes}

upr_genes = Vupr_genes@IntersectionSets %>% 
  tibble( venn = names(.), symbol = .) %>% 
  filter(map_int(symbol ,length)> 0) %>% 
  mutate(symbol = map(symbol, ~ tibble(symbol = .))) %>% 
  unnest() %>% 
  inner_join(diff_genes,by = "symbol")

down_genes = Vdown_genes@IntersectionSets %>% 
  tibble( venn = names(.), symbol = .) %>% 
  filter(map_int(symbol ,length)> 0) %>% 
  mutate(symbol = map(symbol, ~ tibble(symbol = .))) %>% 
  unnest() %>% 
  inner_join(diff_genes,by = "symbol")

```

    - Upregulated

```{r log2fc_upr,fig.width=9,fig.height=4}

upr_genes %>% 
  filter(venn %in% c("0110","0100","0010")) %>% 
  filter(gene_set != "all") %>% 
  nest(-venn) %>% 
  mutate(
    log2fc = map(data,select,symbol,gene_set,log2FoldChange) %>% 
      map(spread,gene_set,log2FoldChange) %>% 
      map(mutate, gene_id = row_number(desc(drdz))) %>% 
      map(select,symbol,gene_id,everything()) %>% 
      map(gather,cell,log2FC,-(1:2))
  ) %>% 
  select(venn,log2fc) %>% 
  unnest() %>% 
  mutate(
    venn = recode_factor(venn,
                  `0100` = "Only EBV",
                  `0010` = "Only NOKS",
                  `0110` = "Both NOKS and EBV")
  ) %>% 
  ggplot(
    aes(gene_id,log2FC,colour = cell)
  )+
  geom_line()+
  facet_grid( ~ venn,scales = "free_x",space = "free_x")+
  theme(
    axis.title.x = element_blank(),
    legend.position = "top"
  )+
  geom_hline(yintercept = 0,linetype = 2)+
  scale_color_brewer(palette = "Set1")


  

```

    - Upregulated

```{r log2fc_down,fig.width=9,fig.height=4}

down_genes %>% 
  filter(venn %in% c("0110","0100","0010")) %>% 
  filter(gene_set != "all") %>% 
  nest(-venn) %>% 
  mutate(
    log2fc = map(data,select,symbol,gene_set,log2FoldChange) %>% 
      map(spread,gene_set,log2FoldChange) %>% 
      map(mutate, gene_id = row_number(desc(drdz))) %>% 
      map(select,symbol,gene_id,everything()) %>% 
      map(gather,cell,log2FC,-(1:2))
  ) %>% 
  select(venn,log2fc) %>% 
  unnest() %>% 
  mutate(
    venn = recode_factor(venn,
                  `0100` = "Only EBV",
                  `0010` = "Only NOKS",
                  `0110` = "Both NOKS and EBV")
  ) %>% 
  ggplot(
    aes(gene_id,log2FC,colour = cell)
  )+
  geom_line()+
  facet_grid( ~ venn,scales = "free_x",space = "free_x")+
  theme(
    axis.title.x = element_blank(),
    legend.position = "top"
  )+
  geom_hline(yintercept = 0,linetype = 2)+
  scale_color_brewer(palette = "Set1")


  

```