---
title: "R Notebook"
output: 
  html_notebook:
    toc: yes
    theme: flatly
---

# Introduction

From Eric's email:

> What the figure aims to show is that the effect of EBV infection is lost with the dRdZ mutant which looks more like an uninfected NOK with respect to the change in the signature genes upon MC treatment.

We prepare the data, basically load and remove genes with very low counts, 
```{r prepare_data,message=FALSE,warning=FALSE}

library(tidyverse)
library(DESeq2)
library(here)
library(Vennerable)
library(gridExtra)
library(ggrepel)

theme_set(theme_bw())

rsem_dr = here("data/RSEM/hg19/Sept17")
align_dr = here("manuscript/logs/Sept17")

## alignment = read_tsv("figs/fig2/Scott_aligned_reads_table.tsv")

stri_subset = function(string,pattern)string[negate(str_detect)(string,pattern)]

alignment = align_dr %>% 
  list.files(full.names = TRUE) %>% 
  stri_subset(".sh") %>% 
  tibble(
    file = .
  ) %>% 
  mutate(
    stats = map(file,read_tsv,col_names = FALSE))

parse_stats = function(stats)
{
  stats %>% 
    filter( 
      str_detect(X1,"total") | str_detect(X1,"mapped")) %>% 
    filter(
      negate(str_detect)(X1,"mate")
    ) %>% 
    mutate(
      type = if_else(str_detect(X1,"total"),"total","aligned"),
      reads = str_split(X1," ") %>% 
        map_chr( ~ .[1]) %>% 
        as.numeric()) %>% 
    select(-X1) 
}

alignment = alignment %>% 
  mutate(
    stats = map(stats,parse_stats),
    file = basename(file)
  ) %>% 
  unnest() %>% 
  spread(type,reads) %>% 
  mutate(
    perc = aligned / total,
    file = str_replace(file,".logs","")
  ) %>% 
  mutate(
    treatment = file %>% 
      str_detect("methyl") %>% 
      if_else(
        "methyl","none"
      ),
    cell = case_when(
          str_detect(file,"clone") ~ "dRdZ",
          str_detect(file,"akata") ~ "EBV" ,
          TRUE ~ "NOKS"),
    rep = file %>% 
      str_split("-") %>% 
      map_chr( ~ .[length(.)]) %>% 
      str_replace("clone","") %>% 
      str_replace("rep","") %>% 
      as.integer())


rsem_data = tibble(
  file = list.files(rsem_dr,full.names = TRUE)) %>% 
  mutate(
    rsem = map(file,read_tsv,progress = FALSE)
  ) %>% 
  mutate(
    file = basename(file) %>% 
      str_replace(".genes.results","")
  ) %>% 
  inner_join(
    select_if(alignment,negate(is.list)), by ="file"
  ) %>% 
  select(-aligned,-total,-perc)

## remove contaminated samples clone2 and clone4
alignment = alignment %>% 
  filter(!( cell == "dRdZ" & rep %in% c(2,4) ))

rsem_data = rsem_data %>% 
  filter(!( cell == "dRdZ" & rep %in% c(2,4) ))

rm(align_dr,alignment)

as_matrix <- function(x)
{
  x %>% 
    as.data.frame() %>% 
    tibble::remove_rownames() %>% 
    tibble::column_to_rownames("gene_id") %>% 
    as.matrix()
}

count_matrix = rsem_data %>% 
  select(file,rsem) %>% 
  unnest() %>% 
  select(file,gene_id,expected_count) %>% 
  mutate(
    expected_count = floor(expected_count)
  ) %>% 
  spread(file,expected_count) %>% 
  as_matrix()

coldata = rsem_data %>% 
  select(file,cell,treatment) %>% 
  mutate(interac = paste(cell,treatment, sep = ".")) %>% 
  as.data.frame() %>% 
  tibble::remove_rownames() %>% 
  tibble::column_to_rownames("file") %>% 
  mutate_if(is_character,as.factor)

```

Since we want to obtain the genes that are differentially expressed (upregulated / downregulated) due to the treatment's effect, we need to prepare two models, one coded in terms of the full model and another in terms of the interactions. 

```{r deseq}
deseq = DESeqDataSetFromMatrix(
  count_matrix,colData = coldata,
  design = ~ interac
)

thr = 20

deseq = deseq[ rowSums(assay(deseq) ) > thr,]
deseq_full = deseq
design(deseq_full) = ~ cell + treatment + cell * treatment

deseq = DESeq(deseq,quiet = TRUE)
deseq_full = DESeq(deseq_full,quiet = TRUE)
```


```{r sanity_check}
### Quick sanity check:

identical(sizeFactors(deseq),sizeFactors(deseq_full))

```

We are going to use the following function to test for hypothesis, contrast needs to be a vectors of the type `c("cat","A","B")`, and is going to test `A vs B` in the `cat` category.

```{r contrasts}

do_contrast = function(deseq,contrast)
{
  results(deseq,
          cooksCutoff = FALSE,
          contrast = contrast,
          tidy = TRUE) %>% 
    as_tibble() %>% 
    dplyr::rename(
      gene_id = row
    ) %>% 
    mutate(
      log10pval = -log10(pvalue)
    ) %>% 
    separate(gene_id,into = c("ensembl","symbol"),sep = "\\_",extra = "drop")
  
}

```

1. Get differentially expressed genes for all genes, and by cell

```{r treatment_genes}

all_diff_genes = do_contrast(deseq_full,c("treatment","methyl","none"))
ebv_diff_genes = do_contrast(deseq, c("interac","EBV.methyl","EBV.none"))
noks_diff_genes = do_contrast(deseq, c("interac","NOKS.methyl","NOKS.none"))
drdz_diff_genes = do_contrast(deseq, c("interac","dRdZ.methyl","dRdZ.none"))

```

2. Compare EBV and NOKS differentially expressed genes vs all by `methyl` treat

```{r venn,out.height=2}

symbols_expressed = function(results, fdr, upregulated = c(NA,TRUE,FALSE))
{
  results = results %>% 
    filter(padj <= fdr)
  
  if(!is.na(upregulated)){
    if(upregulated)results = results %>% filter(log2FoldChange > 0)
    else results = results %>% filter(log2FoldChange < 0 )
  }
  
  results %>% 
    pull(symbol)
  
}

fdr = 0.001
upr = TRUE
upr_genes = list(
  "all" = symbols_expressed(all_diff_genes,fdr,upr),
  "ebv" = symbols_expressed(ebv_diff_genes,fdr,upr),
  "noks" = symbols_expressed(noks_diff_genes,fdr,upr)
) %>% Venn()

upr = FALSE
down_genes = list(
  "all" = symbols_expressed(all_diff_genes,fdr,upr),
  "ebv" = symbols_expressed(ebv_diff_genes,fdr,upr),
  "noks" = symbols_expressed(noks_diff_genes,fdr,upr)
) %>% Venn() 



```

  - Upregulated genes


```{r upr_genes}


plot(upr_genes,doWeights = FALSE,
     show = list(Faces = FALSE))



```

  - Downregulated genes
  
```{r down_reg}
plot(down_genes,doWeights = FALSE,
     show = list(Faces = FALSE))

```


3. Compare the genes that are differentially expressed by treatment for ebv or noks but not all.

```{r gene_sets,fig.width=8,fig.height=4}

upr_genes = upr_genes@IntersectionSets %>% 
  tibble(
    label = names(.),
    genes = .
  ) %>% 
  filter( map_int(genes,length) > 0  ) %>% 
  mutate(
    genes = map(genes, ~ tibble(symbol = .))
  ) %>% 
  unnest() %>% 
  inner_join(
    ebv_diff_genes %>% 
      select(symbol,log2FoldChange) %>% 
      dplyr::rename(ebv = log2FoldChange),by = "symbol") %>% 
  inner_join(
    noks_diff_genes %>% 
      select(symbol,log2FoldChange) %>% 
      dplyr::rename(noks = log2FoldChange), by = "symbol") %>% 
  inner_join(
    drdz_diff_genes %>% 
      select(symbol,log2FoldChange) %>% 
      dplyr::rename(drdz = log2FoldChange),by = "symbol")
  
down_genes = down_genes@IntersectionSets %>% 
  tibble(
    label = names(.),
    genes = .
  ) %>% 
  filter( map_int(genes,length) > 0  ) %>% 
  mutate(
    genes = map(genes, ~ tibble(symbol = .))
  ) %>% 
  unnest() %>% 
  inner_join(
    ebv_diff_genes %>% 
      select(symbol,log2FoldChange) %>% 
      dplyr::rename(ebv = log2FoldChange),by = "symbol") %>% 
  inner_join(
    noks_diff_genes %>% 
      select(symbol,log2FoldChange) %>% 
      dplyr::rename(noks = log2FoldChange), by = "symbol") %>% 
  inner_join(
    drdz_diff_genes %>% 
      select(symbol,log2FoldChange) %>% 
      dplyr::rename(drdz = log2FoldChange),by = "symbol")

```




```{r plot_upr,fig.width=9,fig.height=5}

plot_genes = upr_genes %>% 
  mutate(
    drdz_minus_noks = drdz - noks,
    drdz_minus_ebv = drdz - ebv
  ) %>% 
  arrange(desc(drdz_minus_noks)) %>% 
  group_by(label) %>% 
  mutate(
    symbol_id = seq_len(length(symbol))
  ) %>%
  ungroup() %>% 
  select(label,symbol,symbol_id,contains("minus")) %>% 
  filter(str_detect(label,"0[0|1][0|1]"))  %>% 
  gather( Cell,difference, -(1:3)) %>% 
  mutate(
    label = factor(label),
    label = recode(label , 
                   '010' = "EBV only",
                   '001' = "NOKS only",
                   '011' = "Both"),
    Cell = recode(Cell,
                  'ebv' = "EBV",
                  'noks' = "NOKS",
                  "drdz"= "dRdZ")
  ) 

label_genes = plot_genes %>% 
  filter(abs(difference ) > 1) %>% 
  filter(!(label == "NOKS only" & Cell == "drdz_minus_noks" )) %>% 
  filter(!(label == "EBV only" & Cell == "drdz_minus_ebv")) %>% 
  filter(!label == "Both")

plot_genes %>% 
  ggplot(aes(symbol_id,difference ,colour = Cell))+
  geom_line()+
  facet_grid( . ~ label,space = "free",scales = "free")+
  scale_color_brewer(palette= "Set1",name = "Difference")+
  theme(
    legend.position = "top",
    axis.title.x = element_blank()
  )+
  geom_hline(yintercept = 0,linetype =2 )+
#  geom_text_repel(data = label_genes,aes(label = symbol),show.legend = FALSE)+
  ggtitle("Upregulated genes")


```

```{r plot_down,fig.width=9,fig.height=5}

plot_genes = down_genes %>% 
  mutate(
    drdz_minus_noks = drdz - noks,
    drdz_minus_ebv = drdz - ebv
  ) %>% 
  arrange(desc(drdz_minus_noks)) %>% 
  group_by(label) %>% 
  mutate(
    symbol_id = seq_len(length(symbol))
  ) %>%
  ungroup() %>% 
  select(label,symbol,symbol_id,contains("minus")) %>% 
  filter(str_detect(label,"0[0|1][0|1]"))  %>% 
  gather( Cell,difference, -(1:3)) %>% 
  mutate(
    label = factor(label),
    label = recode(label , 
                   '010' = "EBV only",
                   '001' = "NOKS only",
                   '011' = "Both"),
    Cell = recode(Cell,
                  'ebv' = "EBV",
                  'noks' = "NOKS",
                  "drdz"= "dRdZ")
  ) 

label_genes = plot_genes %>% 
  filter(abs(difference ) > 1) %>% 
  filter(!(label == "NOKS only" & Cell == "drdz_minus_noks" )) %>% 
  filter(!(label == "EBV only" & Cell == "drdz_minus_ebv")) %>% 
  filter(!label == "Both")

plot_genes %>% 
  ggplot(aes(symbol_id,difference ,colour = Cell))+
  geom_line()+
  facet_grid( . ~ label,space = "free",scales = "free")+
  scale_color_brewer(palette= "Set1",name = "Difference")+
  theme(
    legend.position = "top",
    axis.title.x = element_blank()
  )+
  geom_hline(yintercept = 0,linetype =2 )+
#  geom_text_repel(data = label_genes,aes(label = symbol),show.legend = FALSE,
#                  min.segment.length = 1)+
  ggtitle("Downregulated genes")


```


