---
title: "Figures 2018-12-26"
output:
  html_document:
    toc: true
    toc_depth: 2
    self_contained: no
    theme: cerulean    
---

# Main figures

```{r setup, include=FALSE,eval=TRUE,echo=FALSE}

knitr::opts_chunk$set(echo = FALSE,eval = TRUE,
                      include = TRUE,warning = FALSE,
                      message = FALSE,
                      fig.width = 5,fig.height = 5,
                      fig.align = "center")

library(here)
library(rwlib)
library(Vennerable)
library(DESeq2)
library(magrittr)
library(tidyverse)
library(scales)
library(ggrepel)
library(gridExtra)
library(conflicted)

conflict_prefer("filter","dplyr")
conflict_prefer("lag","dplyr")
conflict_prefer("extract","tidyr")
conflict_prefer("Position","ggplot2")
conflict_prefer("col_factor","readr")
conflict_prefer("discard","purrr")
conflict_prefer("expand","tidyr")
conflict_prefer("rename","dplyr")
conflict_prefer("desc","dplyr")
conflict_prefer("select","dplyr")



theme_set(
  theme_bw()+
    theme(
      strip.background = element_blank(),
      axis.text = element_text(size =12),
      axis.title = element_text(size = 14),
      legend.text = element_text(size =12),
      legend.title = element_text(size = 14)
      
    ))

```


```{r load_data, include=FALSE}

noks_vs_ebv_NT = here("apps/Fig2_dashboad/fig2data.RData") %>% 
  load2env() %>% 
  as.list() %>% 
  pluck("diff_genes")

scott_data = here("apps/Fig3_dashboad/fig3data.RData") %>% 
  load2env() %>% 
  as.list()

new_data = here("apps/Fig5_dashboard/fig5data.RData") %>% 
  load2env() %>% 
  as.list()


```


```{r plot_rfuns}

gene_expr_fold_change_plot <- function(DT,xvar,yvar,sc = 10,nbins = 50)
{
  pal = viridis::viridis(1e3, option = "D")
  
  ggplot(DT,aes_string(xvar,yvar))+stat_binhex(bins = nbins) +
    scale_fill_gradientn(colours = pal,trans = "log10",
                         labels = trans_format('log10',math_format(10^.x)),
                         guide = guide_colourbar(title = NULL,
                           barheight = unit(0.5,"npc"),
                           barwidth = unit(0.02,"npc")))+
    xlim(-sc,sc)+ylim(-sc,sc)+
    geom_hline(yintercept = 0,linetype = 2, colour = "black")+
    geom_vline(xintercept = 0,linetype = 2,colour = "black")+
    geom_abline(slope = 1,intercept = 0,linetype = 2, colour = "red")
}


```

## Fig 3A 

Mark mentioned that the idea of this figure is to show that there a not many genes being differentially expressed when comparing NOKS vs EBV. I think that unless we add more elements like:

1. Label genes of interest

2. Add lines resembling the fold change values used for fig. 3B (which I am told is one of the criteria used to draw that figure)

It only makes harder for the reader to catch the point of the figure. For that purpose, I propose the figure below which compares the number of differentially expressed genes explicitly.


```{r noks_vs_ebv_volcano_plot}

fdr = 1e-3

noks_vs_ebv_NT %<>% 
  mutate(
    diff_expr = if_else(padj <= fdr,"yes","no")
  )

noks_vs_ebv_NT %>% 
  ggplot(aes(log2FoldChange,-log10(pvalue),colour = diff_expr))+
  geom_point(alpha = .5,size =2)+
  labs(
    x = expression(log[2]('fold change')),
    y = expression(-log[10]('p.value')),
    color = paste("Diff. Expressed",
              paste0("(FDR <= ",fdr,")"),sep = "\n")
  )+
  theme(
    legend.position = "top"
  )+
  scale_color_brewer(palette = "Set1")+
  scale_x_continuous(limits = c(-10,10))+
  scale_y_continuous(limits = c(0,25))

```

### Alternative


```{r nr_diff_expr_genes, fig.width=3,fig.height=2.8}

noks_vs_ebv_NT %>% 
  group_by(diff_expr) %>% tally() %>% 
  ggplot(aes(x =  diff_expr))+
  geom_col(aes(y = n),width = .5)+
  geom_text(aes(y = n ,label = comma(n)), vjust= -.5)+
  theme(
    axis.title.x = element_blank(),
    legend.position = "none"
  )+
  labs(
    y = paste("Number of diff. expr. genes",
              paste0("(FDR <= ",fdr,")"),sep = "\n")
  )+
  scale_y_continuous(labels = comma_format(accuracy = 1e3),
                     limits = c(0,16.8e3))

```

## Figure 4 

I feel the same way about this figure as I do about the previous one. I am not completely sure about the argument here, but I think that probably we would like to see a Venn diagram comparing the 3 sets of differentially expressed genes or at least these two.

```{r NT_vs_MC_volcano_plots,fig.align="default",fig.width=4,fig.height=4}

scott_data$diff_genes_NOKS %<>%
    mutate(diff_expr = if_else(padj <= fdr,"yes","no")) 

scott_data$diff_genes_EBV %<>%
    mutate(diff_expr = if_else(padj <= fdr,"yes","no")) 


scott_data$diff_genes_NOKS %>% 
  ggplot(aes(log2FoldChange,log10pval,colour = diff_expr))+
  geom_point(alpha = .5,size =2)+
  labs(
    x = expression(log[2]('fold change')),
    y = expression(-log[10]('p.value')),
    color = paste("Diff. Expressed",
              paste0("(FDR <= ",fdr,")"),sep = "\n"),
    subtitle = "NOKs (NT vs. MC)",
    title = "A"
  )+
  theme(
    legend.position = "top",
    plot.subtitle = element_text(hjust = .5, size = 12),
    plot.title = element_text(size = 14)
  )+
  scale_color_brewer(palette = "Set1")+
  scale_x_continuous(limits = c(-10,10))+
  scale_y_continuous(limits = c(0,150))

scott_data$diff_genes_EBV %>% 
  mutate(diff_expr = if_else(padj <= fdr,"yes","no")) %>% 
  ggplot(aes(log2FoldChange,log10pval,colour = diff_expr))+
  geom_point(alpha = .5,size =2)+
  labs(
    x = expression(log[2]('fold change')),
    y = expression(-log[10]('p.value')),
    color = paste("Diff. Expressed",
              paste0("(FDR <= ",fdr,")"),sep = "\n"),
    subtitle = "NOKs-AKATA (NT vs. MC)",
    title = "B"
  )+
  theme(
    legend.position = "top",
    plot.subtitle = element_text(hjust = .5, size = 12),
    plot.title = element_text(size = 14)
  )+
  scale_color_brewer(palette = "Set1")+
  scale_x_continuous(limits = c(-10,10))+
  scale_y_continuous(limits = c(0,150))




```



### Alternative


```{r NT_vs_MC_barplots,fig.height=2.8,fig.width=4.5}

NT_vs_MC_number_genes = 
  tribble(
    ~ cell, ~ genes,
    "NOKs",scott_data$diff_genes_NOKS,
    "NOKs-AKATA", scott_data$diff_genes_EBV
  ) %>% 
  mutate(
    genes = map(genes,group_by,diff_expr) %>% 
      map(tally)
  ) %>% 
  unnest()

NT_vs_MC_number_genes %>% 
  ggplot(aes(x =  diff_expr))+
  geom_col(aes(y = n),width = .5)+
  geom_text(aes(y = n,label = comma(n)), vjust= -.5)+
  theme(
    axis.title.x = element_blank(),
    legend.position = "none"
  )+
  labs(
    y = paste("Number of diff. expr. genes",
              paste0("(FDR <= ",fdr,")"),sep = "\n")
  )+
  scale_y_continuous(labels = comma_format(accuracy = 1e3),
                     limits = c(0,15e3))+
  facet_grid( ~ cell)
```



```{r NT_vs_MC_venn,eval=FALSE}

nt_vs_mc_venn = tribble(
    ~ cell, ~ genes,
    "NOKs",scott_data$diff_genes_NOKS,
    "NOKs-AKATA", scott_data$diff_genes_EBV
  ) %>% 
  unnest() %>% 
  filter(padj <= fdr) %>% 
  select(cell,gene_id) %>% 
  {split(.,.$cell)} %>% 
  map(pluck,"gene_id") %>% 
  Venn()

plot(nt_vs_mc_venn, show = list(Faces =  FALSE))

```


```{r NT_vs_MC_upregulated_venn, eval=FALSE}

nt_vs_mc_venn = tribble(
    ~ cell, ~ genes,
    "NOKs",scott_data$diff_genes_NOKS,
    "NOKs-AKATA", scott_data$diff_genes_EBV
  ) %>% 
  unnest() %>% 
  filter(padj <= fdr & log2FoldChange > 0) %>% 
  select(cell,gene_id) %>% 
  {split(.,.$cell)} %>% 
  map(pluck,"gene_id") %>% 
  Venn()

plot(nt_vs_mc_venn,show = list(Faces =  FALSE))


```

```{r NT_vs_MC_downregulated_venn, eval=FALSE}

nt_vs_mc_venn = tribble(
    ~ cell, ~ genes,
    "NOKs",scott_data$diff_genes_NOKS,
    "NOKs-AKATA", scott_data$diff_genes_EBV
  ) %>% 
  unnest() %>% 
  filter(padj <= fdr & log2FoldChange < 0) %>% 
  select(cell,gene_id) %>% 
  {split(.,.$cell)} %>% 
  map(pluck,"gene_id") %>% 
  Venn()

plot(nt_vs_mc_venn,show = list(Faces =  FALSE))


```

## Figure 5

### Gene expression analysis

To make this figure, we need to make some preparations. First, we need to find the genes that are differentially expressed by either the treatment or the cell lines.

```{r scott_genes_diff , include=FALSE}

as_matrix <- function(x)
{
  x %>% 
    as.data.frame() %>% 
    tibble::remove_rownames() %>% 
    tibble::column_to_rownames("gene_id") %>% 
    as.matrix()
}

count_matrix = scott_data %>% 
  pluck("rsem_data") %>% 
  select(file,rsem) %>% 
  unnest() %>% 
  select(file,gene_id,expected_count) %>% 
  mutate(
    expected_count = floor(expected_count)
  ) %>% 
  spread(file,expected_count) %>% 
  as_matrix()

coldata = scott_data %>% 
  pluck("rsem_data") %>% 
  select(file,cell,treatment) %>% 
  mutate(interac = paste(cell,treatment, sep = ".")) %>% 
  as.data.frame() %>% 
  tibble::remove_rownames() %>% 
  tibble::column_to_rownames("file")

deseq = DESeqDataSetFromMatrix(
  count_matrix,colData = coldata,
  design = ~ interac
)

## remove genes with low read counts, filtered genes with no reads but probably can 
## filter genes with low expression

thr = 20

deseq = deseq[ rowSums(assay(deseq) ) > thr,]
deseq = DESeq(deseq)

scott_data$diff_genes = results(
  deseq,
  cooksCutoff = FALSE,tidy = TRUE) %>% 
  as_tibble() %>% 
  dplyr::rename(
    gene_id = row
  ) %>% 
  mutate(   ### changing the fold change and stat to be MC vs none
    log2FoldChange = - log2FoldChange,
    stat = -stat,           
    log10pval = -log10(pvalue)
  )

nt_vs_mc = tribble(
    ~ cell, ~ genes,
    "Both",scott_data %>% pluck("diff_genes"),
    "NOKs",scott_data %>% pluck("diff_genes_NOKS"),
    "NOKs-AKATA", scott_data %>% pluck("diff_genes_EBV")
  )

```

Since we were doing the volcano plots, it would be useful to examine it for this set of genes too.

```{r volcano_default , fig.width=4,fig.height=4, warning=FALSE}

scott_data$diff_genes %>% 
  mutate(diff_expr = if_else(padj <= fdr,"yes","no")) %>% 
  ggplot(aes(log2FoldChange,log10pval,colour = diff_expr))+
  geom_point(alpha = .5,size =2)+
  labs(
    x = expression(log[2]('fold change')),
    y = expression(-log[10]('p.value')),
    color = paste("Diff. Expressed",
              paste0("(FDR <= ",fdr,")"),sep = "\n")
  )+
  theme(
    legend.position = "top",
    plot.subtitle = element_text(hjust = .5, size = 12),
    plot.title = element_text(size = 14)
  )+
  scale_color_brewer(palette = "Set1")+
  scale_x_continuous(limits = c(-10,10))+
  scale_y_continuous(limits = c(0,150))
```

Under the same argument as before, I think this could be an alternative to the figure above:

```{r nr_scott_diff_genes,fig.width=3,fig.height=2.8}

scott_data %>% 
  pluck("diff_genes") %>% 
  mutate(
    diff_expr = if_else(padj <= fdr,"yes","no")
  ) %>% 
  group_by(diff_expr) %>% tally() %>% 
  ggplot(aes(x =  diff_expr))+
  geom_col(aes(y = n),width = .5)+
  geom_text(aes(y = n ,label = comma(n)), vjust= -.5)+
  theme(
    axis.title.x = element_blank(),
    legend.position = "none"
  )+
  labs(
    y = paste("Number of diff. expr. genes",
              paste0("(FDR <= ",fdr,")"),sep = "\n")
  )+
  scale_y_continuous(labels = comma_format(accuracy = 1e3),
                     limits = c(0,12e3))


```

### Venn diagram {.tabset .tabset-fade}

The Venn diagrams below were constructed by defining a genes to be differentially expressed iff the adjust p.values were below `r fdr`. First, we compared all genes and noticed that a lot of genes are being differentially expressed regardless of the cell line. However, when the genes are divided by the regulation sign ( i.e. the $\log_2(\mbox{fold change}))$ sign) we can notice that the same pattern holds, thus an apparent conclusion is that there is not a pattern respect to this sign. I think we should use this to motivate the pathway analysis.

I feel it is a bit odd that there are so many genes being differentially expressed with both cell lines, only NOKs, or only NOKs-AKATA. How are those genes shared?

In the Venn diagrams below, we consider the genes that are differentially expressed MC vs. NT:

- Both: Means, when both cell lines are considered in the test
- NOKs and NOKs-AKATA: Only one of those cell lines


```{r plot_venn_color,include=FALSE,eval=TRUE}

library(grid)


plot_venn <- function(venn , colors)
{
  gp = VennThemes(venn)
  
  gp[["Face"]][[1]]$fill = "white"
  gp[["Face"]][[2]]$fill = "white"
  gp[["Face"]][[3]]$fill = "white"
  gp[["Face"]][[4]]$fill = "white"
  gp[["Face"]][[5]]$fill = "white"
  gp[["Face"]][[6]]$fill = "white"
  gp[["Face"]][[7]]$fill = "white"
  gp[["Face"]][[8]]$fill = "white"
  
  
  gp[["Set"]][[1]]$col = colors[1]
  gp[["Set"]][[2]]$col = colors[2]
  gp[["Set"]][[3]]$col = colors[3]

  gp[["SetText"]][[1]]$col = colors[1]
  gp[["SetText"]][[2]]$col = colors[2]
  gp[["SetText"]][[3]]$col = colors[3]

  
  grid.newpage()
  plot(venn,gp = gp)  
}



```


#### All diff expressed genes

```{r venn_all}

nt_vs_mc_venn = nt_vs_mc %>% 
  unnest() %>% 
  filter(padj <= fdr) %>% 
  select(cell,gene_id) %>% 
  {split(.,.$cell)} %>% 
  map(pluck,"gene_id") %>% 
  Venn() %>% 
  compute.Venn(doWeights = FALSE)

plot_venn(nt_vs_mc_venn,
          colors = c("#5380E4", "#E12A3C", "#FFBF03")) 

```

#### Upregulated

```{r venn_upregulated}

nt_vs_mc_venn = nt_vs_mc %>% 
  unnest() %>% 
  filter(padj <= fdr & log2FoldChange > 0) %>% 
  select(cell,gene_id) %>% 
  {split(.,.$cell)} %>% 
  map(pluck,"gene_id") %>% 
  Venn() %>% 
  compute.Venn(doWeights = FALSE)

plot_venn(nt_vs_mc_venn,
          colors = c("#5380E4", "#E12A3C", "#FFBF03")) 

```

#### Downregulated

```{r venn_downregulated}

nt_vs_mc_venn = nt_vs_mc %>% 
  unnest() %>% 
  filter(padj <= fdr & log2FoldChange < 0) %>% 
  select(cell,gene_id) %>% 
  {split(.,.$cell)} %>% 
  map(pluck,"gene_id") %>% 
  Venn() %>% 
  compute.Venn(doWeights = FALSE)

plot_venn(nt_vs_mc_venn,
          colors = c("#5380E4", "#E12A3C", "#FFBF03")) 

```

### Pathway analysis {.tabset .tabset-fade}

I had a bit of trouble generating the plot below using bars, so I thought to use points as it looks cleaner. I added the Venn diagrams as possible plots for the supplement. Something that I found interesting, although not surprising, is that the Venn diagrams are a bit similar at both gene and pathway levels.

```{r pathway_preparation}

library(clusterProfiler)

gmt_files =  here::here("data/Pathway") %>% 
  list.files(full.names = TRUE,recursive = TRUE) %>% 
  stri_subset("Intere") %>% 
  stri_subset("kera") %>% 
  set_names(
    {
      c("hallmark")
    }
  )

gene_lists = tibble(
 pathway_name = names(gmt_files), files = gmt_files
) %>% 
  mutate(
    pathways = files %>% 
      map(read.gmt) %>% map(as_tibble)
  ) %>% 
  dplyr::select(-files)

name <- function(.x){
  nms = pull(.x,symbol)
  pull(.x,stat) %>% set_names(nms)
}

nt_vs_mc %<>%
  mutate(
    results = map(genes,separate,gene_id,into = c("ensembl","symbol"),sep = "_")) %>% 
  mutate(
    all_genes_list = map(results,arrange,desc(stat) ) %>% map(name),
    diff_genes_list = map(results,filter,padj <= fdr) %>% 
      map(arrange,desc(stat)) %>% map(name))


nt_vs_mc %<>% crossing(.,gene_lists)

nt_vs_mc %<>%
  mutate(
    all_genes_list_gsea = map2( all_genes_list, pathways,
                                ~ GSEA(.x,TERM2GENE = .y,verbose = FALSE)),
    diff_genes_list_gsea = map2( diff_genes_list, pathways,
                                ~ GSEA(.x,TERM2GENE = .y,verbose = FALSE)))

nt_vs_mc %<>% 
  mutate_at(vars(contains("gsea")),
            funs( res =  map(. , "result") %>% 
                    map(as_tibble)))
```


```{r pathway_all_genes, fig.width= 10,fig.height=9}

library(ggtech)

nt_vs_mc %>% 
  select(cell,all_genes_list_gsea_res) %>% 
  unnest() %>% 
  group_by(ID) %>% 
  summarize(NES = if_else(all(cell == "Both"), 0, median(NES[cell != "Both"]))) %>% 
  ungroup() %>% 
  arrange(NES) %>% 
  pull(ID) -> all_ids

nt_vs_mc %>% 
  select(cell,all_genes_list_gsea_res) %>% 
  unnest() %>% 
  mutate(
    only_cell = if_else(str_detect(cell,"NOK"),1,2),
    ID = factor(ID,levels = all_ids)
  ) %>% 
  ggplot(aes(x = NES,y = ID , color = cell))+
  geom_point(size = 2)+
  theme(legend.position = "top",
        axis.title.y = element_blank(),
        strip.background = element_blank(),
        strip.text = element_blank())+
  scale_color_tech(theme = "google")+
  facet_grid(~ only_cell)+
  geom_vline(xintercept = 0,linetype = 2)
  

```


#### All pathways

```{r venn_pathways_all}

nt_vs_mc %>% 
  select(cell,all_genes_list_gsea_res) %>% 
  unnest() %>% 
  select(cell,ID) %>% 
  {split(.,.$cell)} %>% 
  map(pluck,"ID") %>% 
  Venn() %>% 
  compute.Venn(doWeights = FALSE)-> pathway_venn


plot_venn(pathway_venn,
          colors = c("#5380E4", "#E12A3C", "#FFBF03")) 


```

#### Upregulated pathways

```{r venn_pathways_up}

nt_vs_mc %>% 
  select(cell,all_genes_list_gsea_res) %>% 
  unnest() %>% 
  filter(NES > 0 ) %>% 
  select(cell,ID) %>% 
  {split(.,.$cell)} %>% 
  map(pluck,"ID") %>% 
  Venn() %>% 
  compute.Venn(doWeights = FALSE)-> pathway_venn


plot_venn(pathway_venn,
          colors = c("#5380E4", "#E12A3C", "#FFBF03")) 


```

#### Downregulated pathways

```{r venn_pathways_down}

nt_vs_mc %>% 
  select(cell,all_genes_list_gsea_res) %>% 
  unnest() %>% 
  filter(NES < 0) %>% 
  select(cell,ID) %>% 
  {split(.,.$cell)} %>% 
  map(pluck,"ID") %>% 
  Venn() %>% 
  compute.Venn(doWeights = FALSE)-> pathway_venn

plot_venn(pathway_venn,
          colors = c("#5380E4", "#E12A3C", "#FFBF03")) 


```

### Differentiated genes pathway analysis {.tabset .tabset-fade}

If we only consider the genes that are differentially expressed (under our definition p.value < `r fdr`), it seems to be almost the same. If anything the number of meaningful pathways seems to be reduced, producing a more conservative pathway analysis (i.e. probably a bit of signal of non-differentially expressed genes is filtering). However, it is not surprising because the GSEA test works by examining the tails of the signal to noise ratio, which are almost the same between both analysis because the number of differentially expressed genes is very high.

```{r pathway_diff_genes, fig.width= 10,fig.height=9}

library(ggtech)

nt_vs_mc %>% 
  select(cell,diff_genes_list_gsea_res) %>% 
  unnest() %>% 
  group_by(ID) %>% 
  summarize(NES = if_else(all(cell == "Both"), 0, median(NES[cell != "Both"]))) %>% 
  ungroup() %>% 
  arrange(NES) %>% 
  pull(ID) -> all_ids

nt_vs_mc %>% 
  select(cell,diff_genes_list_gsea_res) %>% 
  unnest() %>% 
  mutate(
    only_cell = if_else(str_detect(cell,"NOK"),1,2),
    ID = factor(ID,levels = all_ids)
  ) %>% 
  ggplot(aes(x = NES,y = ID , color = cell))+
  geom_point(size = 2)+
  theme(legend.position = "top",
        axis.title.y = element_blank(),
        strip.background = element_blank(),
        strip.text = element_blank())+
  scale_color_tech(theme = "google")+
  facet_grid(~ only_cell)+
  geom_vline(xintercept = 0,linetype = 2)
  

```

#### All pathways

```{r diff_venn_pathways_all}

nt_vs_mc %>% 
  select(cell,diff_genes_list_gsea_res) %>% 
  unnest() %>% 
  select(cell,ID) %>% 
  {split(.,.$cell)} %>% 
  map(pluck,"ID") %>% 
  Venn() %>% 
  compute.Venn(doWeights = FALSE)-> pathway_venn


plot_venn(pathway_venn,
          colors = c("#5380E4", "#E12A3C", "#FFBF03")) 


```

#### Upregulated pathways

```{r diff_venn_pathways_up}

nt_vs_mc %>% 
  select(cell,diff_genes_list_gsea_res) %>% 
  unnest() %>% 
  filter(NES > 0 ) %>% 
  select(cell,ID) %>% 
  {split(.,.$cell)} %>% 
  map(pluck,"ID") %>% 
  Venn() %>% 
  compute.Venn(doWeights = FALSE)-> pathway_venn

plot_venn(pathway_venn,
          colors = c("#5380E4", "#E12A3C", "#FFBF03")) 


```

#### Downregulated pathways

```{r diff_venn_pathways_down}

nt_vs_mc %>% 
  select(cell,diff_genes_list_gsea_res) %>% 
  unnest() %>% 
  filter(NES < 0) %>% 
  select(cell,ID) %>% 
  {split(.,.$cell)} %>% 
  map(pluck,"ID") %>% 
  Venn() %>% 
  compute.Venn(doWeights = FALSE)-> pathway_venn


plot_venn(pathway_venn,
          colors = c("#5380E4", "#E12A3C", "#FFBF03")) 


```


## Figure 6A

__Scott Fold change:__ The idea of this figure is two fold. a) Compare the MC over NT fold change between NOKS and EBV cell lines. b) Quantify the ratio between cell fold changes. 

For b), we considered the $y  = \beta x + \epsilon$ model, where $x$ and $y$ are the $\log_2 \left(\frac{1 + \mbox{mean MC TPM}}{mean NT TPM}\right)$ for the NOKs, and NOKs-AKATA samples respectively. Below, we calculated the $H_0: \beta = 1$ test, representing that 




```{r scott_fold_change, results='hold'}

scott_fc = scott_data$rsem_data %>% 
  unnest() %>% 
  mutate(treatment = if_else(treatment == "MC","MC","NT")) %>% 
  select(file,cell,treatment,gene_id,TPM) %>% 
  group_by(gene_id,cell,treatment) %>% 
  summarize(
    TPM = mean(TPM)
  ) %>% 
  spread(treatment,TPM) %>% 
  mutate(
    log2fc = log2((1 + MC)/ (1 + NT))
  ) %>% 
  select(-MC,-NT) %>% 
  spread(cell,log2fc)

scott_fc %>% 
  gene_expr_fold_change_plot("NOKS","EBV",sc = 8,nbins = 60)+
  coord_fixed()+
  labs(
    x = "NOKs",
    y = "NOKs-AKATA"
  )+
  geom_smooth(se = FALSE,method = "lm",formula = y ~ 0 + x,size = 1,
              colour = "magenta",linetype = 2,size = .5)

update_geom_defaults("smooth", list(size = .5))

fc_model = lm( EBV ~ 0 + NOKS , data = scott_fc)

multcomp::glht(model = fc_model, linfct = c("NOKS = 1")) %>% 
  summary()


```

### Fig 6B (or supplement)

We repeated the previous analysis, but by keeping only the genes that are differentially expressed by cell or treatment (i.e. the 'Both' group). In magenta, we observe the slope obtained by considering only the differentially expressed genes, and in orange we observe the linear relationship by restricting the genes by their regulation (i.e. NOKs > 0 and NOKs-AKATA > 0 and NOKs < 0 and NOKs-AKATA < 0 ).

```{r scott_fold_change_diff, results='hold'}

scott_fc_diff = scott_fc %>% 
  inner_join(
    scott_data %>% pluck("diff_genes") %>% select(gene_id,log2FoldChange,padj),by = "gene_id"
  )  %>% 
  filter(padj < fdr)
  
scott_fc_diff %>% 
  gene_expr_fold_change_plot("NOKS","EBV",sc = 8,nbins = 60)+
  coord_fixed()+
  labs(
    x = "NOKs",
    y = "NOKs-AKATA"
  )+
  geom_smooth(se = FALSE,method = "lm",formula = y ~ 0 + x,size = 1,
              colour = "magenta",linetype = 2,size = .5)+
  geom_smooth(se = FALSE,method = "lm",formula = y ~ 0 + x,size = 1,
              colour = "orange",linetype = 2,size = .5,
              data = filter(scott_fc_diff, EBV > 0 & NOKS > 0))+
  geom_smooth(se = FALSE,method = "lm",formula = y ~ 0 + x,size = 1,
              colour = "orange",linetype = 2,size = .5,
              data = filter(scott_fc_diff, EBV < 0 & NOKS < 0))

fc_model = lm( EBV ~ 0 + NOKS , data = scott_fc_diff)

multcomp::glht(model = fc_model, linfct = c("NOKS = 1")) %>% 
  summary()


```

In the figure above, the genes appear to be clustered into two groups (i.e. genes that are up and down-regulated), thus, we fit the same linear model but restricting the differentially expressed genes by the regulation sign. This is barely noticeable in the plot

```{r diff_up_model, include=TRUE,eval=TRUE}

fc_model_up = lm( EBV ~ 0 + NOKS , 
                  data = scott_fc_diff %>% filter(EBV > 0 & NOKS > 0))

multcomp::glht(model = fc_model_up, linfct = c("NOKS = 1")) %>% 
  summary()


```
```{r diff_down_model, include=TRUE,eval=TRUE}

fc_model_up = lm( EBV ~ 0 + NOKS , 
                  data = scott_fc_diff %>% filter(EBV < 0 & NOKS < 0 ))

multcomp::glht(model = fc_model_up, linfct = c("NOKS = 1")) %>% 
  summary()


```

First, we can notice that the when the genes are down-regulated the ratio between NOKs-AKATA and NOKS fold change is more similar to the ratio obtained when considering all the genes (0.75), than  the ratio when considering only the differentially expressed genes (0.79). The table below, illustrates the number of genes in each regulation block:

```{r ngenes_table,include=TRUE,eval=TRUE}

scott_fc_diff %>% 
  mutate(
    regulation = case_when(
      EBV > 0 & NOKS > 0 ~ "Both up" ,
      EBV < 0 & NOKS < 0 ~ "Both down",
      EBV > 0 & NOKS < 0 ~ "EBV up and NOKS down",
      EBV < 0 & NOKS > 0 ~ "EBV down and NOKS up"
    ) ) %>% 
  group_by(regulation) %>% tally() %>% 
  knitr::kable(digits = 0)

```


# Supplement figures

## Supp. Fig 1

__Scott data QC:__ We removed the EBV - NT first replicate due to lower quality.

```{r scott_aligned_reads}

scott_data$alignment %>% 
  mutate(
    treat = if_else(treat == "None", "NT",treat) %>% 
      factor() %>% fct_rev(),
    cell = factor(cell) %>% fct_rev()
  ) %>% 
  ggplot(aes(aligned_percent,aligned))+
  geom_point()+
  facet_grid(
    treat ~ cell
  )+
  geom_text_repel(aes(label = replicate))+
  scale_x_continuous(labels = percent_format(accuracy = 1))+
  scale_y_continuous(labels = comma_format(scale = 1e-6,suffix = "M"))+
  labs(
    x = "Percentage of aligned reads",
    y = "Number of aligned reads"
  )

```

<!-- ## Supp. Fig 2 -->

<!-- __New data QC:__ We removed replicates 2 and 4 of $\Delta R \Delta Z$ because they were contaminated. -->

```{r new_aligned_reads,fig.width = 7.5,include=FALSE,eval=FALSE}

new_data$alignment %>%
  rename(treat = treatment) %>% 
  mutate(
    treat = if_else(treat == "none", "NT","MC") %>%
      factor() %>% fct_rev(),
    cell = factor(cell) %>% fct_rev(),
    rep = paste0("rep",rep)
  ) %>%
  ggplot(aes(perc,aligned))+
  geom_point()+
  facet_grid(
    treat ~ cell
  )+
  geom_text_repel(aes(label = rep))+
  scale_x_continuous(labels = percent_format(accuracy = 1))+
  scale_y_continuous(labels = comma_format(scale = 1e-6,suffix = "M"))+
  labs(
    x = "Percentage of aligned reads",
    y = "Number of aligned reads"
  )

```

## Supp. Fig 2



```{r pvalue_hist_scott,fig.width=4,fig.height=3,fig.align="default"}

scott_data$diff_genes %>% 
  ggplot(aes(pvalue))+
  geom_histogram(colour = "black",fill = "white",boundary = 0)+
  labs(
    title = "A",
    subtitle = "NT vs. MC"
  )

scott_data$diff_genes_EBV %>% 
  ggplot(aes(pvalue))+
  geom_histogram(colour = "black",fill = "white",boundary = 0)+
  labs(
    title = "B",
    subtitle = "NOKs-AKATA: NT vs. MC"
  )

scott_data$diff_genes_NOKS %>% 
  ggplot(aes(pvalue))+
  geom_histogram(colour = "black",fill = "white",boundary = 0)+
  labs(
    title = "C",
    subtitle = "NOKs: NT vs. MC"
    )


```

