<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Figures 2018-12-26</title>

<script src="Figures_paper_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="Figures_paper_files/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="Figures_paper_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="Figures_paper_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="Figures_paper_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="Figures_paper_files/navigation-1.1/tabsets.js"></script>
<link href="Figures_paper_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="Figures_paper_files/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>



<div class="container-fluid main-container">

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->





<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Figures 2018-12-26</h1>

</div>

<div id="TOC">
<ul>
<li><a href="#main-figures">Main figures</a><ul>
<li><a href="#fig-3a">Fig 3A</a></li>
<li><a href="#figure-4">Figure 4</a></li>
<li><a href="#figure-5">Figure 5</a></li>
</ul></li>
<li><a href="#supplement-figures">Supplement figures</a><ul>
<li><a href="#supp.-fig-1">Supp. Fig 1</a></li>
<li><a href="#supp.-fig-2">Supp. Fig 2</a></li>
<li><a href="#supp.-fig-3">Supp. Fig 3</a></li>
</ul></li>
</ul>
</div>

<div id="main-figures" class="section level1">
<h1>Main figures</h1>
<div id="fig-3a" class="section level2">
<h2>Fig 3A</h2>
<p>Mark mentioned that the idea of this figure is to show that there a not many genes being differentially expressed when comparing NOKS vs EBV. I think that unless we add more elements like:</p>
<ol style="list-style-type: decimal">
<li><p>Label genes of interest</p></li>
<li><p>Add lines resembling the fold change values used for fig. 3B (which I am told is one of the criteria used to draw that figure)</p></li>
</ol>
<p>It only makes hardes for the reader to catch the point of the figure. For that purpose, I propose the figure below which compares the number of differentially expressed genes explicitely.</p>
<p><img src="Figures_paper_files/figure-html/noks_vs_ebv_volcano_plot-1.png" width="480" style="display: block; margin: auto;" /></p>
<div id="alternative" class="section level3">
<h3>Alternative</h3>
<p><img src="Figures_paper_files/figure-html/nr_diff_expr_genes-1.png" width="288" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="figure-4" class="section level2">
<h2>Figure 4</h2>
<p>I feel the same way about this figure as I do about the previous one. I am not completely sure about the argument here, but I think that probably we would like to see a Venn diagram comparing the 3 sets of differentially expressed genes or at least these two.</p>
<p><img src="Figures_paper_files/figure-html/NT_vs_MC_volcano_plots-1.png" width="384" /><img src="Figures_paper_files/figure-html/NT_vs_MC_volcano_plots-2.png" width="384" /></p>
<div id="alternative-1" class="section level3">
<h3>Alternative</h3>
<p><img src="Figures_paper_files/figure-html/NT_vs_MC_barplots-1.png" width="432" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="figure-5" class="section level2">
<h2>Figure 5</h2>
<div id="gene-expression-analysis" class="section level3">
<h3>Gene expression analysis</h3>
<p>To make this figure, we need to make some preparations. First, we need to find the genes that are differentially expressed by either the treatment or the cell lines.</p>
<p>Since we were doing the volcano plots, it would be useful to examine it for this set of genes too.</p>
<p><img src="Figures_paper_files/figure-html/volcano_default%20-1.png" width="384" style="display: block; margin: auto;" /></p>
<p><img src="Figures_paper_files/figure-html/nr_scott_diff_genes-1.png" width="288" style="display: block; margin: auto;" /></p>
</div>
<div id="venn-diagram" class="section level3 tabset tabset-fade">
<h3>Venn diagram</h3>
<p>The venn diagrams below were constructed by defining a genes to be differentially expressed iff the adjust p.values were below 0.001. First, we compared all genes and noticed that a lot of genes are being differentially expressed regardless of the cell line. However, when the genes are divided by the regulation sign ( i.e. the <span class="math inline">\(\log_2(\mbox{fold change}))\)</span> sign) we can notice that the same pattern holds, thus an apparent conclusion is that there is not a pattern respect to this sign. I think we should use this to motivate the pathway analysis.</p>
<p>I feel it is a bit odd that there are so many genes being differentially expressed with both cell lines, only NOKs, or only NOKs-AKATA. How are those genes shared?</p>
<p>In the venn diagrams below, we consider the genes that are differentially expressed MC vs. NT:</p>
<ul>
<li>Both: Means, when both cell lines are considered in the test</li>
<li>NOKs and NOKs-AKATA: Only one of those cell lines</li>
</ul>
<div id="all-diff-expressed-genes" class="section level4">
<h4>All diff expressed genes</h4>
<p><img src="Figures_paper_files/figure-html/venn_all-1.png" width="480" style="display: block; margin: auto;" /></p>
</div>
<div id="upregulated" class="section level4">
<h4>Upregulated</h4>
<p><img src="Figures_paper_files/figure-html/venn_upregulated-1.png" width="480" style="display: block; margin: auto;" /></p>
</div>
<div id="downregulated" class="section level4">
<h4>Downregulated</h4>
<p><img src="Figures_paper_files/figure-html/venn_downregulated-1.png" width="480" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="pathway-analysis" class="section level3 tabset tabset-fade">
<h3>Pathway analysis</h3>
<p><img src="Figures_paper_files/figure-html/pathway_all_genes-1.png" width="960" style="display: block; margin: auto;" /></p>
<div id="all-pathways" class="section level4">
<h4>All pathways</h4>
<p><img src="Figures_paper_files/figure-html/venn_pathways_all-1.png" width="480" style="display: block; margin: auto;" /></p>
</div>
<div id="upregulated-pathways" class="section level4">
<h4>Upregulated pathways</h4>
<p><img src="Figures_paper_files/figure-html/venn_pathways_up-1.png" width="480" style="display: block; margin: auto;" /></p>
</div>
<div id="downregulated-pathways" class="section level4">
<h4>Downregulated pathways</h4>
<p><img src="Figures_paper_files/figure-html/venn_pathways_down-1.png" width="480" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="differentiated-genes-pathway-analysis" class="section level3 tabset tabset-fade">
<h3>Differentiated genes pathway analysis</h3>
<p>If we only consider the genes that are differentially expressed (under our definition p.value &lt; 0.001), it seems to be almost the same. If anything the number of meaningful pathways seems to be reduced, producing a more conservative pathway analysis (i.e. probably a bit of signal of non-differentially expressed genes is filtering).</p>
<p><img src="Figures_paper_files/figure-html/pathway_diff_genes-1.png" width="960" style="display: block; margin: auto;" /></p>
<div id="all-pathways-1" class="section level4">
<h4>All pathways</h4>
<p><img src="Figures_paper_files/figure-html/diff_venn_pathways_all-1.png" width="480" style="display: block; margin: auto;" /></p>
</div>
<div id="upregulated-pathways-1" class="section level4">
<h4>Upregulated pathways</h4>
<p><img src="Figures_paper_files/figure-html/diff_venn_pathways_up-1.png" width="480" style="display: block; margin: auto;" /></p>
</div>
<div id="downregulated-pathways-1" class="section level4">
<h4>Downregulated pathways</h4>
<p><img src="Figures_paper_files/figure-html/diff_venn_pathways_down-1.png" width="480" style="display: block; margin: auto;" /> ## Figure 6A</p>
<p><strong>Scott Fold change:</strong> The idea of this figure is two fold. a) Compare the MC over NT fold change between NOKS and EBV cell lines. b) If possible quantify</p>
<pre><code>## 
##   Simultaneous Tests for General Linear Hypotheses
## 
## Fit: lm(formula = EBV ~ 0 + NOKS, data = scott_fc)
## 
## Linear Hypotheses:
##              Estimate Std. Error  t value Pr(&gt;|t|)    
## NOKS == 1    0.750509   0.002312 -107.927 2.22e-16 ***
## NOKS == 0.77 0.750509   0.002312   -8.432  &lt; 2e-16 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## (Adjusted p values reported -- single-step method)</code></pre>
<p><img src="Figures_paper_files/figure-html/scott_fold_change-1.png" width="480" style="display: block; margin: auto;" /></p>
</div>
</div>
</div>
</div>
<div id="supplement-figures" class="section level1">
<h1>Supplement figures</h1>
<div id="supp.-fig-1" class="section level2">
<h2>Supp. Fig 1</h2>
<p><strong>Scott data QC:</strong> We removed the EBV - NT first replicate due to lower quality.</p>
<p><img src="Figures_paper_files/figure-html/scott_aligned_reads-1.png" width="480" style="display: block; margin: auto;" /></p>
</div>
<div id="supp.-fig-2" class="section level2">
<h2>Supp. Fig 2</h2>
<p><strong>New data QC:</strong> We removed replicates 2 and 4 of <span class="math inline">\(\Delta R \Delta Z\)</span> because they were contaminated.</p>
<p><img src="Figures_paper_files/figure-html/new_aligned_reads-1.png" width="720" style="display: block; margin: auto;" /></p>
</div>
<div id="supp.-fig-3" class="section level2">
<h2>Supp. Fig 3</h2>
<p><img src="Figures_paper_files/figure-html/pvalue_hist_scott-1.png" width="384" /><img src="Figures_paper_files/figure-html/pvalue_hist_scott-2.png" width="384" /></p>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
