---
title: "EBV and NOKS: Treated (MC) vs Untreated"
output:
  flexdashboard::flex_dashboard:
    theme: yeti
runtime: shiny
---

```{r global,include=TRUE,echo = FALSE,eval = TRUE}

load("fig3data.RData")

## block to check if the packages are installed
list.of.packages <- c("tidyverse","ggrepel","viridis","scales","shiny",
                      "rmarkdown","flexdashboard","d3heatmap",
                      "shinythemes","DT","RColorBrewer")
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)

library(tidyverse)
library(ggrepel)
library(viridis)
library(scales)
library(shiny)
library(d3heatmap)

my_alpha = .5
pal = viridis(1000)

theme_set(
  theme_bw()+
    theme(legend.position = "top",
          axis.text = element_text(size = 14),
          axis.title = element_text(size = 16))
)

common_genes = diff_genes_EBV %>% 
  select(gene_id) %>% 
  inner_join(
    diff_genes_NOKS %>% select(gene_id), by = "gene_id"
  )

NOKS_vs_EBV_log2FC = rsem_data %>% 
  select(cell,treatment,rsem) %>%
  unnest() %>% 
  nest(-cell,-treatment) %>% 
  mutate(
    data = map(data,select,gene_id,TPM) %>% 
      map(group_by,gene_id) %>% 
      map(summarize,
          TPM = mean(TPM))
  ) %>% 
  unnest() %>% 
  nest(-cell)%>% 
  mutate(
    data = map(data,spread,treatment,TPM) %>% 
      map(mutate, log2FC = log2(1 + MC) - log2(1 + none)) %>% 
      map(select,gene_id,log2FC)
  ) %>% 
  unnest() %>% 
  spread(cell,log2FC) 

# 
# NOKS_vs_EBV_log2FC %>% 
#   ggplot(aes(NOKS,EBV))+stat_binhex(bins = 101) +
#   scale_fill_gradientn(colours = pal,trans = "log10",
#                          labels = trans_format('log10',math_format(10^.x)),
#                          guide = guide_colourbar(title = NULL,
#                            barheight = unit(0.92,"npc"),
#                            barwidth = unit(0.01,"npc")))+
#   theme(legend.position = "right")
# 

MA_TPM = rsem_data %>% 
  select(cell,treatment,rsem) %>% 
  unnest() %>% 
  group_by(cell,treatment,gene_id) %>% 
  summarize(
    TPM = mean(TPM)
  ) %>% 
  ungroup() %>% 
  nest(-cell,.key = cell_MA) %>% 
  mutate(
    cell_MA = map(cell_MA,spread,treatment,TPM)
  ) %>% 
  unnest() %>% 
  mutate_if(is.numeric,funs(log2(1 + .))) %>% 
  mutate(
    A = 0.5 * (MC + none),
    M = MC - none
  ) %>% 
  nest(-cell)

diff_genes = 
  tibble(
    cell = c("EBV","NOKS"),
    results = list(diff_genes_EBV,diff_genes_NOKS)
  ) %>% 
  inner_join(
    MA_TPM,by = "cell"
  ) %>% 
  mutate(
    results = map2(results,data, inner_join,by  = "gene_id"),
    data = NULL
  ) %>% 
  unnest() %>% 
  separate(gene_id,into = c("ensembl","gene_id"),sep = "\\_") %>% 
  nest(-cell,.key = results)

stri_subset = function(string,pattern)
{
  string[! str_detect(string,pattern)]
}

tpmmat = rsem_data %>% 
  select(file,rsem) %>% 
  unnest() %>% 
  select(file,gene_id,TPM) %>% 
  spread(file,TPM) %>% 
  as.data.frame() %>% 
  tibble::remove_rownames() %>% 
  tibble::column_to_rownames("gene_id") %>% 
  as.matrix()

```
    
Inputs {.sidebar}
-------------------------------------

```{r sidebar}
# shiny inputs defined here

  numericInput("fdr","False Discovery Rate",1e-20,
               min = 1e-6,max = 1,step = .01)
  numericInput("ps","Point size",4,
               min = .5,max = 20,step = .5)
  
  hr()
  
  actionButton("clean","Clean Gene List")
  
  radioButtons("heatmap_type", "Heatmap type type",
            c("most DE", "selected"), inline = TRUE)
  
  selectizeInput("genes_used","Genes used",
                 choices = c("All",pluck(diff_genes,"cell")))
  
  numericInput("heatmap_most","Most DE genes",10,
               min = 5,max = 100,step = 1)
  
  hr()
  
  actionButton("match","Match gene lists")
  
  hr()
  
  actionButton("save","Save genes")
  textInput("file",label = "Genes file", value = "fig3_genes.tsv")
  
  ## fix these
  observeEvent(input$save,{
    if(!is.null(my_genes$ensembl)){
      to_save = diff_genes %>% 
        filter(ensembl %in% my_genes$ensembl)
      
      write_tsv(to_save,input$file)
      
    }
  })
  
  my_genes <- reactiveValues(ensembl = NULL)
  ebv_genes <- reactiveValues(ensembl = NULL)
  noks_genes <- reactiveValues(ensembl = NULL)
  
  EBV_genes = diff_genes %>% 
    filter(cell == "EBV") %>% 
    pluck("results") %>% 
    pluck(1)
  
  NOKS_genes = diff_genes %>% 
    filter(cell == "NOKS") %>% 
    pluck("results") %>% 
    pluck(1)
  
  gene_colors = c("black","darkblue","firebrick4")    
  #   RColorBrewer::brewer.pal(3,"Dark2")[1:2]
  ## action button events  
  #  ntext <- eventReactive(input$goButton, {
  #   input$n
  # })
  # 
  # output$nText <- renderText({
  #   ntext()
  # })
  
  observeEvent(input$clean,{
    ebv_genes$ensembl <- NULL
    noks_genes$ensembl <- NULL
  })
  
```

   
Column {.tabset}
-------------------------------------
   
### Alignment

```{r alignment}

alignment %>% 
    ggplot(
        aes(aligned,aligned_percent))+
    geom_point()+
    geom_text_repel(aes(label = replicate))+
    facet_grid(treat ~ cell)+
    theme(axis.text.x = element_text(angle = 25,
                                     hjust = 1))+
    scale_x_continuous(
        labels = comma)+
    scale_y_continuous(
        labels = percent)+
    xlab("Number of aligned reads")+
    ylab("Percentage of aligned reads")
```   

### p.value

```{r pval_histogram}

diff_genes %>% 
  unnest() %>% 
  ggplot(aes(pvalue))+
  geom_histogram(boundary = 0,
                 fill = "white",
                 color = "black",
                 bins = 51)+
  facet_grid( cell ~ .)+
  ylab("Number of genes")

```

### NOKS vs EBV 

### Volcano EBV 
    
```{r volcano_ebv}

volcano_ebv_ranges <- reactiveValues(x = NULL,y = NULL)

observeEvent(input$volcano_ebv_dblclick, {
  brush <- input$volcano_ebv_brush
  if (!is.null(brush)) {
    volcano_ebv_ranges$x <- c(brush$xmin, brush$xmax)
    volcano_ebv_ranges$y <- c(brush$ymin, brush$ymax)
    volcano_ebv_ranges$y[1] <- max(volcano_ebv_ranges$y[1],0)
  } else {
    volcano_ebv_ranges$x <- NULL
    volcano_ebv_ranges$y <- NULL
  }
})

observeEvent(input$volcano_ebv_click,{
  ff <- nearPoints(EBV_genes,input$volcano_ebv_click)
  ebv_genes$ensembl <- c(ebv_genes$ensembl, ff$ensembl)
})


output$volcano_ebv = renderPlot({

  both = intersect(ebv_genes$ensembl,
                   noks_genes$ensembl)
  only_ebv = setdiff(ebv_genes$ensembl,
                     noks_genes$ensembl)
  only_noks = setdiff(noks_genes$ensembl,
                      ebv_genes$ensembl)
  
  both_data = EBV_genes %>% 
    filter(ensembl %in% both)
  
  ebv_data = EBV_genes %>%
    filter(ensembl %in% only_ebv)
  
  noks_data = EBV_genes %>% 
    filter(ensembl %in% only_noks)

  EBV_genes %>%
    ggplot(aes(log2FoldChange,log10pval,
               colour = if_else(padj <= input$fdr,"yes","no")))+
    geom_point(size = input$ps,alpha = .75)+
    geom_text_repel(data = both_data,aes(label = gene_id),
                    colour = gene_colors[1],
                    size = 4)+
    geom_text_repel(data = ebv_data,aes(label = gene_id),
                    colour = gene_colors[2],
                    size = 4)+
    geom_text_repel(data = noks_data,aes(label = gene_id),
                    colour = gene_colors[3],
                    size = 4)+
    scale_color_brewer(palette = "Set1",name = "Diff. Expressed")+
    coord_cartesian(
      xlim = volcano_ebv_ranges$x ,
      ylim = volcano_ebv_ranges$y
    )+
    ylab("-log10(pvalue)")+
    geom_vline(xintercept = 0,colour = "black",linetype = 2)
})
 
fillCol(flex = c(1,NA),
        plotOutput("volcano_ebv",
                   click = "volcano_ebv_click",
                   dblclick = "volcano_ebv_dblclick",
                   brush = brushOpts(
                     id = "volcano_ebv_brush",
                     resetOnNew = TRUE
                     ))
        )


```

### Volcano NOKS

```{r volcano_noks}

volcano_noks_ranges <- reactiveValues(x = NULL,y = NULL)

observeEvent(input$volcano_noks_dblclick, {
  brush <- input$volcano_noks_brush
  if (!is.null(brush)) {
    volcano_noks_ranges$x <- c(brush$xmin, brush$xmax)
    volcano_noks_ranges$y <- c(brush$ymin, brush$ymax)
    volcano_noks_ranges$y[1] <- max(volcano_noks_ranges$y[1],0)
  } else {
    volcano_noks_ranges$x <- NULL
    volcano_noks_ranges$y <- NULL
  }
})

observeEvent(input$volcano_noks_click,{
  ff <- nearPoints(NOKS_genes,input$volcano_noks_click)
  noks_genes$ensembl <- c(noks_genes$ensembl, ff$ensembl)
})


output$volcano_noks = renderPlot({

  both = intersect(ebv_genes$ensembl,
                   noks_genes$ensembl)
  only_ebv = setdiff(ebv_genes$ensembl,
                     noks_genes$ensembl)
  only_noks = setdiff(noks_genes$ensembl,
                      ebv_genes$ensembl)
  
  both_data = NOKS_genes %>% 
    filter(ensembl %in% both)
  ebv_data = NOKS_genes %>%
    filter(ensembl %in% only_ebv)
  noks_data = NOKS_genes %>% 
    filter(ensembl %in% only_noks)

  NOKS_genes %>%
    ggplot(aes(log2FoldChange,log10pval,
               colour = if_else(padj <= input$fdr,"yes","no")))+
    geom_point(size = input$ps,alpha = .75)+
    geom_text_repel(data = both_data,aes(label = gene_id),
                    colour = gene_colors[1],
                    size = 4)+
    geom_text_repel(data = ebv_data,aes(label = gene_id),
                    colour = gene_colors[2],
                    size = 4)+
    geom_text_repel(data = noks_data,aes(label = gene_id),
                    colour = gene_colors[3],
                    size = 4)+
    scale_color_brewer(palette = "Set1",name = "Diff. Expressed")+
    coord_cartesian(
      xlim = volcano_noks_ranges$x ,
      ylim = volcano_noks_ranges$y
    )+
    ylab("-log10(pvalue)")+
    geom_vline(xintercept = 0,colour = "black",linetype = 2)
})
 
fillCol(flex = c(1,NA),
        plotOutput("volcano_noks",
                   click = "volcano_noks_click",
                   dblclick = "volcano_noks_dblclick",
                   brush = brushOpts(
                     id = "volcano_noks_brush",
                     resetOnNew = TRUE
                     ))
        )


```

### MA EBV

```{r , ma_ebv}

# MA2_ranges <- reactiveValues(x = NULL, y = NULL)
# 
# observeEvent(input$MA2_dblclick, {
#   brush <- input$MA2_brush
#   if (!is.null(brush)) {
#     MA2_ranges$x <- c(brush$xmin, brush$xmax)
#     MA2_ranges$x[1] <- max(0 , MA1_ranges$x[1])
#     MA2_ranges$y <- c(brush$ymin, brush$ymax)
#   } else {
#     MA2_ranges$x <- NULL
#     MA2_ranges$y <- NULL
#   }
# })
# 
# observeEvent(input$MA2_click,{
#   ff <- nearPoints(diff_genes,input$MA2_click)
#   my_genes$ensembl <- c(my_genes$ensembl, ff$ensembl)
# })
# 
# output$MA2 = renderPlot({
#   
#   gene_data = diff_genes %>% 
#     filter(ensembl %in% my_genes$ensembl)
#   
#   diff_genes %>% 
#     ggplot(aes(A,M))+
#     scale_color_brewer(palette = "Set1", name = "Diff. Expressed")+
#     geom_point(aes(colour = if_else(padj <= input$fdr,"yes","no")),
#                alpha = .75,size = input$ps)+
#     geom_text_repel(data = gene_data,aes(label = gene_id),colour = "black",
#                     size = 4)+
#     coord_cartesian(
#       xlim = MA2_ranges$x,
#       ylim = MA2_ranges$y
#     )+
#     geom_hline(yintercept = 0,colour = "black",linetype = 2)
# })
# 
# fillCol(flex = c(1,NA),
#         plotOutput("MA2",
#                    click = "MA2_click",
#                    dblclick = "MA2_dblclick",
#                    brush = brushOpts(
#                      id = "MA2_brush",
#                      resetOnNew = TRUE
#                      ))
#         )
# 

```

### MA NOKS

### Model heatmap EBV

```{r model_heatmap}

# renderD3heatmap({
#   
#   my_mat = rlogmat1
#   
#   if(input$heatmap_type == "most DE"){
#     genes = diff_genes %>% 
#       arrange(desc(log10pval)) %>% 
#       head(input$heatmap_most )
#     
#       idx = genes %>% 
#         mutate(
#           idx = paste(ensembl,gene_id,sep = "_")
#         ) %>% 
#         pluck("idx")
#       my_mat = my_mat[rownames(my_mat) %in% idx, ]
#     
#   }else{
#     if(is.null(my_genes$ensembl)){
#       my_mat = my_mat[sample(nrow(my_mat),3),]
#     }else{
#       genes = diff_genes %>% 
#         filter(ensembl %in% my_genes$ensembl)
#     
#       idx = genes %>% 
#         mutate(
#           idx = paste(ensembl,gene_id,sep = "_")
#         ) %>% 
#         pluck("idx")
#       my_mat = my_mat[rownames(my_mat) %in% idx, ]
#     }
#     
#   }
#   rownames(my_mat) = rownames(my_mat) %>% 
#     str_split("_") %>% 
#     map_chr( ~ .[2])
#   
#   colnames(my_mat) = colnames(my_mat) %>% 
#     str_replace("RNAseq-","")
#   
#   pal = viridis(100)
#   
#   d3heatmap(my_mat,
#             colors = pal,
#             xaxis_font_size = 12,
#             yaxis_font_size = 12)
# })


```

### Model Heatmap NOKS

### TPM heatmap

```{r tpm_heatmap}

# renderD3heatmap({
#   
#   my_mat = tpmmat1
#   
#   if(input$heatmap_type == "most DE"){
#     genes = diff_genes %>% 
#       arrange(desc(log10pval)) %>% 
#       head(input$heatmap_most )
#     
#       idx = genes %>% 
#         mutate(
#           idx = paste(ensembl,gene_id,sep = "_")
#         ) %>% 
#         pluck("idx")
#       my_mat = my_mat[rownames(my_mat) %in% idx, ]
#     
#   }else{
#     if(is.null(my_genes$ensembl)){
#       my_mat = my_mat[sample(nrow(my_mat),3),]
#     }else{
#       genes = diff_genes %>% 
#         filter(ensembl %in% my_genes$ensembl)
#     
#       idx = genes %>% 
#         mutate(
#           idx = paste(ensembl,gene_id,sep = "_")
#         ) %>% 
#         pluck("idx")
#       my_mat = my_mat[rownames(my_mat) %in% idx, ]
#     }
#     
#   }
#   rownames(my_mat) = rownames(my_mat) %>% 
#     str_split("_") %>% 
#     map_chr( ~ .[2])
#   
#   colnames(my_mat) = colnames(my_mat) %>% 
#     str_replace("RNAseq-","")
#   
#   pal = viridis(100)
#   
#   d3heatmap(my_mat,
#             scale = "row",
#             colors = pal,
#             xaxis_font_size = 12,
#             yaxis_font_size = 12)
# })

```

### Genes table

```{r table}  
##actionLink("generate","Generate gene list!")

# div(
#   renderDataTable(
#     diff_genes %>% 
#       filter(ensembl %in% my_genes$ensembl) %>% 
#       select(gene_id,ensembl,padj,log10pval,EBV,NOKS,stat),
#     options = list(
#       pageLength = 10,
#       lengthMenu = c(10,20,50,100)
#     ) ) ,style = "font-size:80%")

```

### Instructions

**Alignment** - This plot is fixed, I think for the supplement

**p.values** - Also fixed, there is no deviation of the assumptions. Both look like they are supposed: a peak at zero and then approx. uniform. A quick note, is that the peak at zero is really tall, which means that there are a lot of diff. expressed genes. We may want to use smaller values for setting the FDR (by default 1e-20).

**volcano** and **MA plots** - In these plots, can click the point and build a table of DE genes. To change the scale of the figure: brush and double click. If you double click without brush that will restart to the initial plot. Just click on a gene to label it, the results are stored in both **heatmaps** and **Gene table**. Removed the MA-DESeq plot because it didn't seem very useful.

**Sidebar**

* False discovery rate: The smaller, the test results are more conservative

* Point size: Just to make the plots to look just right

* Clean gene lists: To start over again (changed to have one per cell)

* Save genes and file: When some genes are selected from the volcano or MA plots,this button will save the list in the file



