---
title: "Untreated samples: NOKS+EBV vs NOKS"
output:
  flexdashboard::flex_dashboard:
    theme: yeti
runtime: shiny
---

```{r global,include=TRUE,echo = FALSE,eval = TRUE}

load("fig2data.RData")

## block to check if the packages are installed
list.of.packages <- c("tidyverse","ggrepel","viridis","scales","shiny",
                      "rmarkdown","flexdashboard","d3heatmap",
                      "shinythemes","DT","RColorBrewer","extrafontdb","pheatmap")
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages))
  suppressMessages(install.packages(new.packages,repos = "http://cran.rstudio.com"))

library(tidyverse)
library(ggrepel)
library(viridis)
library(scales)
library(shiny)
library(d3heatmap)

my_alpha = .5
my_style = "color: #FFFFFF; background-color: #FF0000"
my_style2 = "color: #FFFFFF; background-color: #357EC7"

theme_set(
  theme_bw()+
    theme(legend.position = "top",
          axis.text = element_text(size = 14),
          axis.title = element_text(size = 16))
)

MA_tpm = rsem_data %>% 
  filter(treatment == "none") %>% 
  select(file,cell,rsem) %>% 
  unnest() %>% 
  group_by(cell,gene_id) %>% 
  summarize(
    tpm = mean(TPM)
  ) %>% 
  ungroup() %>% 
  spread(cell,tpm) %>% 
  mutate_if(
    is.numeric,
    funs(log2(1 + .))
  ) %>% 
  mutate(
    A = 0.5 * (EBV + NOKS),
    M = EBV - NOKS
  )

diff_genes = inner_join(
  diff_genes,
  MA_tpm,by = "gene_id"
) %>% 
  separate(gene_id,into = c("ensembl","gene_id"),sep = "\\_") %>% 
  mutate(
    log10pval = -log10(pvalue)
  )

stri_subset = function(string,pattern)
{
  string[! str_detect(string,pattern)]
}

rlogmat1 = rlogmat[, colnames(rlogmat) %>% 
                    stri_subset("MC")]

tpmmat1 = rsem_data %>% 
  filter(cell == "NOKS") %>% 
  select(file,rsem) %>% 
  unnest() %>% 
  select(file,gene_id,TPM) %>% 
  spread(file,TPM) %>% 
  as.data.frame() %>% 
  tibble::remove_rownames() %>% 
  tibble::column_to_rownames("gene_id") %>% 
  as.matrix()

```
    
Inputs {.sidebar}
-------------------------------------

```{r sidebar}
# shiny inputs defined here

  numericInput("fdr","False Discovery Rate",0.05,
               min = 1e-6,max = 1,step = .01)
  numericInput("ps","Point size",4)
  numericInput("ts","Axis ticks size",12)
  numericInput("Ts","Axis title size",16)
  numericInput("ls","Label size",6)
  hr()
  
  radioButtons("heatmap_type", "Heatmap type type",
            c("most DE", "selected"), inline = TRUE)
  
  numericInput("heatmap_most","Most DE genes",10)
  
  hr()
  
  textInput("file",label = "Genes file", value = "fig2_genes.tsv")

  hr()
  textInput("loadfile",label = "Load genes",value = "fig2_genes.tsv")

  fillRow(flex = c(.4,.4,.4),
          actionButton("load","Load",icon = icon("folder-open-o"),style = my_style2),
          actionButton("save"," Save",icon = icon("floppy-o"),
                       style = my_style2),
          actionButton("clean"," Clean",icon = icon("window-close"),
                       style = my_style2))

  observeEvent(input$save,{
    if(!is.null(my_genes$ensembl)){
      to_save = diff_genes %>% 
        filter(ensembl %in% my_genes$ensembl)
      
      write_tsv(to_save,input$file)
      
    }
  })
  
  observeEvent(input$clean,{
    my_genes$ensembl <- NULL
  })
  
  observeEvent(input$load,{
    
    if(file.exists(input$loadfile)){
      diff_genes = read_tsv(input$loadfile)
      my_genes$ensembl = pluck(diff_genes,"ensembl")
    }
    
  })
  
  
  
```

   
Column {.tabset}
-------------------------------------
   
### Alignment

```{r alignment}

alignment %>% 
    ggplot(
        aes(aligned,aligned_percent))+
    geom_point()+
    geom_text_repel(aes(label = replicate))+
    facet_grid(treat ~ cell)+
    theme(axis.text.x = element_text(angle = 25,
                                     hjust = 1))+
    scale_x_continuous(
        labels = comma)+
    scale_y_continuous(
        labels = percent)+
    xlab("Number of aligned reads")+
    ylab("Percentage of aligned reads")
```   

### p.value

```{r pval_histogram}

diff_genes %>% 
    ggplot(aes(pvalue))+
    geom_histogram(aes(y = ..count../sum(..count..) ),
                   boundary = 0 , 
                   fill = "white",
                   color = "black",
                 bins = 51)+
  geom_hline(yintercept = 0.015,linetype = 2,colour = "red")

```

### Volcano
    
```{r volcano}

volcano_ranges <- reactiveValues(x = NULL, y = NULL)
my_genes <- reactiveValues(ensembl = NULL)

observeEvent(input$volcano_dblclick, {
  brush <- input$volcano_brush
  if (!is.null(brush)) {
    volcano_ranges$x <- c(brush$xmin, brush$xmax)
    volcano_ranges$y <- c(brush$ymin, brush$ymax)
    volcano_ranges$y[1] <- max(volcano_ranges$y[1],0)
  } else {
    volcano_ranges$x <- NULL
    volcano_ranges$y <- NULL
  }
})


observeEvent(input$volcano_click,{
  ff <- nearPoints(diff_genes,input$volcano_click)
  my_genes$ensembl <- c(my_genes$ensembl, ff$ensembl)
})

volcano = reactive({
  gene_data = diff_genes %>% 
    filter(ensembl %in% my_genes$ensembl)

  diff_genes %>%
    ggplot(aes(log2FoldChange,log10pval))+
    geom_point(size = input$ps,alpha = .75,
               aes(colour = if_else(padj <= input$fdr,"yes","no")))+
    geom_label_repel(data = gene_data,
                  colour = 'white', fill = "black",
                    box.padding = 0.35, point.padding = 0.5,
                    segment.color = 'grey50',
                    aes(label = gene_id),size = input$ls)+
    scale_color_brewer(palette = "Set1",name = "Diff. Expressed")+
    coord_cartesian(
      xlim = volcano_ranges$x ,
      ylim = volcano_ranges$y
    )+
    theme(
      axis.text = element_text(size = input$ts),
      axis.title = element_text(size = input$Ts),
      legend.text = element_text(size = input$ts),
      legend.title = element_text(size = input$ts)
    )+
    ylab("-log10(pvalue)")+
    geom_vline(xintercept = 0,colour = "black",linetype = 2)
})

output$volcano = renderPlot(volcano())

fillCol(flex = c(1,.1),
        plotOutput("volcano",
                   click = "volcano_click",
                   dblclick = "volcano_dblclick",
                   brush = brushOpts(
                     id = "volcano_brush",
                     resetOnNew = TRUE
                     )),
        fillRow(flex = c(.1,.3,.3,.3),
          actionButton("volcano_save","",icon = icon("camera"),
                       style = my_style),
          numericInput("volcano_width","Volcano width (in)",7),
          numericInput("volcano_height","Volcano height (in)",7),
          textInput("volcano_file","Volcano file","volcano_file.pdf")))

observeEvent(input$volcano_save,{

  ggsave(
    plot = volcano(),
    filename = input$volcano_file,
    width = unit(input$volcano_width,"in"),
    height = unit(input$volcano_height, "in")
    
  )

})


```

### MA - DESeq

```{r ma_deseq}

MA1_ranges <- reactiveValues(x = NULL, y = NULL)

observeEvent(input$MA1_dblclick, {
  brush <- input$MA1_brush
  if (!is.null(brush)) {
    MA1_ranges$x <- c(brush$xmin, brush$xmax)
    MA1_ranges$x[1] <- max(0 , MA1_ranges$x[1])
    MA1_ranges$y <- c(brush$ymin, brush$ymax)
  } else {
    MA1_ranges$x <- NULL
    MA1_ranges$y <- NULL
  }
})

observeEvent(input$MA1_click,{
  ff <- nearPoints(diff_genes,input$MA1_click)
  my_genes$ensembl <- c(my_genes$ensembl, ff$ensembl)
})

MA1 = reactive({
  gene_data = diff_genes %>% 
    filter(ensembl %in% my_genes$ensembl)
  
  diff_genes %>% 
    ggplot(aes(baseMean,log2FoldChange))+
    geom_point(size = input$ps,alpha = .75,
               aes(colour = if_else(padj <= input$fdr,"yes","no")))+
    scale_color_brewer(palette = "Set1",name = "Diff. Expressed")+
    geom_label_repel(data = gene_data,
                  colour = 'white', fill = "black",
                    box.padding = 0.35, point.padding = 0.5,
                    segment.color = 'grey50',
                    aes(label = gene_id),size = input$ls)+
    scale_color_brewer(palette = "Set1",name = "Diff. Expressed")+
    coord_cartesian(
      xlim = MA1_ranges$x ,
      ylim = MA1_ranges$y
    )+
    theme(
      axis.text = element_text(size = input$ts),
      axis.title = element_text(size = input$Ts),
      legend.text = element_text(size = input$ts),
      legend.title = element_text(size = input$ts)
    )+
    geom_hline(yintercept = 0,colour = "black",linetype = 2)
  
})

output$MA1 = renderPlot(MA1())

fillCol(flex = c(1,.1),
        plotOutput("MA1",
                   click = "MA1_click",
                   dblclick = "MA1_dblclick",
                   brush = brushOpts(
                     id = "MA1_brush",
                     resetOnNew = TRUE
                   )),
        fillRow(flex = c(.1,.3,.3,.3),
          actionButton("ma1_save","",icon = icon("camera"),
                       style = my_style),
          numericInput("ma1_width","MA width (in)",7),
          numericInput("ma1_height","MA height (in)",7),
          textInput("ma1_file","MA-plot file","ma1_file.pdf")))

observeEvent(input$ma1_save,{

  ggsave(
    plot = MA1(),
    filename = input$ma1_file,
    width = unit(input$ma1_width,"in"),
    height = unit(input$ma1_height, "in")
    
  )

})


```

### MA - data

```{r , ma_tpm}

MA2_ranges <- reactiveValues(x = NULL, y = NULL)

observeEvent(input$MA2_dblclick, {
  brush <- input$MA2_brush
  if (!is.null(brush)) {
    MA2_ranges$x <- c(brush$xmin, brush$xmax)
    MA2_ranges$x[1] <- max(0 , MA1_ranges$x[1])
    MA2_ranges$y <- c(brush$ymin, brush$ymax)
  } else {
    MA2_ranges$x <- NULL
    MA2_ranges$y <- NULL
  }
})

observeEvent(input$MA2_click,{
  ff <- nearPoints(diff_genes,input$MA2_click)
  my_genes$ensembl <- c(my_genes$ensembl, ff$ensembl)
})

MA2 = reactive({
    gene_data = diff_genes %>% 
    filter(ensembl %in% my_genes$ensembl)
  
  diff_genes %>% 
    ggplot(aes(A,M))+
    geom_point(size = input$ps,alpha = .75,
               aes(colour = if_else(padj <= input$fdr,"yes","no")))+
    scale_color_brewer(palette = "Set1",name = "Diff. Expressed")+
    geom_label_repel(data = gene_data,
                  colour = 'white', fill = "black",
                    box.padding = 0.35, point.padding = 0.5,
                    segment.color = 'grey50',
                    aes(label = gene_id),size = input$ls)+
    scale_color_brewer(palette = "Set1",name = "Diff. Expressed")+
    coord_cartesian(
      xlim = MA2_ranges$x ,
      ylim = MA2_ranges$y
    )+
    theme(
      axis.text = element_text(size = input$ts),
      axis.title = element_text(size = input$Ts),
      legend.text = element_text(size = input$ts),
      legend.title = element_text(size = input$ts)
    )+
    geom_hline(yintercept = 0,colour = "black",linetype = 2)

})

output$MA2 = renderPlot(MA2())

fillCol(flex = c(1,.1),
        plotOutput("MA2",
                   click = "MA2_click",
                   dblclick = "MA2_dblclick",
                   brush = brushOpts(
                     id = "MA2_brush",
                     resetOnNew = TRUE
                     )),
        fillRow(flex = c(.1,.3,.3,.3),
          actionButton("ma2_save","",icon = icon("camera"),
                       style = my_style),
          numericInput("ma2_width","MA width (in)",7),
          numericInput("ma2_height","MA height (in)",7),
          textInput("ma2_file","MA-plot file","ma2_file.pdf")))

observeEvent(input$ma2_save,{

  ggsave(
    plot = MA2(),
    filename = input$ma2_file,
    width = unit(input$ma2_width,"in"),
    height = unit(input$ma2_height, "in")
    
  )

})



```



### Model heatmap

```{r model_heatmap}

model_mat = reactive({
  
  my_mat = rlogmat1
  
  if(input$heatmap_type == "most DE"){
    genes = diff_genes %>% 
      arrange(desc(log10pval)) %>% 
      head(input$heatmap_most )
    
      idx = genes %>% 
        mutate(
          idx = paste(ensembl,gene_id,sep = "_")
        ) %>% 
        pluck("idx")
      if(length(idx) <= 1){
        my_mat = my_mat[1:2,]
      }else{
        my_mat = my_mat[rownames(my_mat) %in% idx, ]
      }
    
  }else{
    if(is.null(my_genes$ensembl)){
      my_mat = my_mat[sample(nrow(my_mat),3),]
    }else{
      genes = diff_genes %>% 
        filter(ensembl %in% my_genes$ensembl)
    
      idx = genes %>% 
        mutate(
          idx = paste(ensembl,gene_id,sep = "_")
        ) %>% 
        pluck("idx")
      my_mat = my_mat[rownames(my_mat) %in% idx, ]
    }
    
  }
  rownames(my_mat) = rownames(my_mat) %>% 
    str_split("_") %>% 
    map_chr( ~ .[2])
  
  colnames(my_mat) = colnames(my_mat) %>% 
    str_replace("RNAseq-","")
  
  my_mat

})


fillCol(flex = c(1,.1),
        renderD3heatmap({
          my_mat = model_mat()
          pal = viridis(100)
          d3heatmap(my_mat,
                    colors = pal,
                    xaxis_font_size = 12,
                    yaxis_font_size = 12)
          }),
        fillRow(flex = c(.1,.3,.3,.3),
                actionButton("model_heat_save","",icon = icon("camera"),
                       style = my_style),
          numericInput("model_heat_width","Model Heatmap width (in)",7),
          numericInput("model_heat_height","Model Heatmap height (in)",7),
          textInput("model_heat_file","Model Heatmap file","mh_file.pdf")))

observeEvent(input$model_heat_save,{

  pal = viridis(100)
  
  pheatmap::pheatmap(
    mat = model_mat(),
    scale = "row",
    color = pal,
    fontsize = input$ts,
    width = input$model_heat_width,
    height = input$model_heat_height,
    filename = input$model_heat_file
  )

})



```

### TPM heatmap

```{r tpm_heatmap}

tpm_mat <- reactive({
  
  my_mat = tpmmat1
  
  if(input$heatmap_type == "most DE"){
    genes = diff_genes %>% 
      arrange(desc(log10pval)) %>% 
      head(input$heatmap_most )
    
      idx = genes %>% 
        mutate(
          idx = paste(ensembl,gene_id,sep = "_")
        ) %>% 
        pluck("idx")

      if(length(idx) <= 1){
        my_mat = my_mat[1:2,]
      }else{
        my_mat = my_mat[rownames(my_mat) %in% idx, ]
      }

    
  }else{
    if(is.null(my_genes$ensembl)){
      my_mat = my_mat[sample(nrow(my_mat),3),]
    }else{
      genes = diff_genes %>% 
        filter(ensembl %in% my_genes$ensembl)
    
      idx = genes %>% 
        mutate(
          idx = paste(ensembl,gene_id,sep = "_")
        ) %>% 
        pluck("idx")
      my_mat = my_mat[rownames(my_mat) %in% idx, ]
    }
    
  }
  rownames(my_mat) = rownames(my_mat) %>% 
    str_split("_") %>% 
    map_chr( ~ .[2])
  
  colnames(my_mat) = colnames(my_mat) %>% 
    str_replace("RNAseq-","")

  my_mat
  
  
})




fillCol(flex = c(1,.1),
        renderD3heatmap({
          
          my_mat = tpm_mat()  
          pal = viridis(100)
  
          d3heatmap(my_mat,
                    scale = "row",
                    colors = pal,
                    xaxis_font_size = 12,
                    yaxis_font_size = 12)}),
        fillRow(flex = c(.1,.3,.3,.3),
                actionButton("tpm_heat_save","",icon = icon("camera"),
                       style = my_style),
          numericInput("tpm_heat_width","TPM Heatmap width (in)",7),
          numericInput("tpm_heat_height","TPM Heatmap height (in)",7),
          textInput("tpm_heat_file","TPM Heatmap file","tpmh_file.pdf")))

observeEvent(input$tpm_heat_save,{

  pal = viridis(100)

  pheatmap::pheatmap(
    mat = tpm_mat(),
    color = pal,
    scale = "row",
    fontsize = input$ts,
    width = input$tpm_heat_width,
    height = input$tpm_heat_height,
    filename = input$tpm_heat_file
  )

})



```

### Genes table

```{r table}  
##actionLink("generate","Generate gene list!")

div(
  renderDataTable(
    diff_genes %>% 
      filter(ensembl %in% my_genes$ensembl) %>% 
      select(gene_id,ensembl,padj,log10pval,EBV,NOKS,stat),
    options = list(
      pageLength = 10,
      lengthMenu = c(10,20,50,100)
    ) ) ,style = "font-size:80%")

```

### Instructions

**Alignment** - This plot is fixed, I think for the supplement

**p.value** - Also fixed, there is a deviation of the multiple testing assumptions

**volcano** and **MA plots** - In these three, can click the point and build a table of DE genes. To change the scale of the figure: brush and double click. If you double click without brush that will restart to the initial plot. Just click on a gene to label it, the results are stored in both **heatmaps** and **Gene table**.

**Sidebar**

* False discovery rate: The smaller, the test results are more conservative

* Point size: Just to make the plots to look just right

* Clean gene list: To start over again

* Save genes and file: When some genes are selected from the volcano or MA plots,this button will save the list in the file



